{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import numpy as np\n",
    "\n",
    "DIR = \"/data/project/Meningioma/02.Align/hg38/Tumor/05.Final_bam/230405_Tumor.chr22.bam\"\n",
    "\n",
    "TUMOR_BAM_PYSAM = pysam.AlignmentFile( DIR , 'rb')\n",
    "BLOOD_BAM_PYSAM = pysam.AlignmentFile( DIR.replace(\"Tumor\", \"Blood\") , 'rb')\n",
    "\n",
    "#1. TUMOR 의 read를 돌기\n",
    "prev_pos = -1\n",
    "cnt = 0\n",
    "li_tumor, li_blood = [], []\n",
    "\n",
    "for read in TUMOR_BAM_PYSAM.fetch():\n",
    "    if (read.reference_name == \"chr22\") & ( read.pos > 29603000) & (read.pos < 29700000):\n",
    "        if read.cigarstring != None:  # 왜 None이 있지?\n",
    "            if (\"S\" in read.cigarstring)  |  (\"H\" in read.cigarstring):\n",
    "                #print(\"Pos : {}\\t{}\\tRead name : {}\\t\".format(read.reference_name, read.pos, read.query_name))\n",
    "                if read.pos != prev_pos:\n",
    "                    cnt = 1\n",
    "                    prev_pos = read.pos\n",
    "                else:\n",
    "                    cnt = cnt + 1\n",
    "\n",
    "        \n",
    "        if cnt  >= 2:\n",
    "            li_tumor.append (prev_pos)\n",
    "            cnt = 0\n",
    "\n",
    "#2. BLOOD 의 read를 돌기\n",
    "prev_pos = -1\n",
    "cnt = 0\n",
    "\n",
    "for read in BLOOD_BAM_PYSAM.fetch():\n",
    "    if (read.reference_name == \"chr22\") & ( read.pos > 2960300) & (read.pos < 29700000):\n",
    "        if read.cigarstring != None:  # 왜 None이 있지?\n",
    "            if (\"S\" in read.cigarstring)  |  (\"H\" in read.cigarstring):\n",
    "                #print(\"Pos : {}\\t{}\\tRead name : {}\\t\".format(read.reference_name, read.pos, read.query_name))\n",
    "                if read.pos != prev_pos:\n",
    "                    cnt = 1\n",
    "                    prev_pos = read.pos\n",
    "                else:\n",
    "                    cnt = cnt + 1\n",
    "\n",
    "        \n",
    "        if cnt  >= 2:\n",
    "            li_blood.append (prev_pos)\n",
    "            cnt = 0\n",
    "\n",
    "\n",
    "# Close the BAM file\n",
    "TUMOR_BAM_PYSAM.close()\n",
    "BLOOD_BAM_PYSAM.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "221021\n",
      "INDEL >= 2 :  {29636686: 2, 29636922: 2, 29642294: 2, 29642318: 3, 29661208: 2, 29661237: 2, 29678244: 2}\n",
      "221028\n",
      "INDEL >= 2 :  {29639209: 2, 29658329: 2, 29665059: 2, 29668381: 3, 29668366: 2, 29683037: 6, 29683075: 6, 29683093: 6}\n",
      "221202\n",
      "INDEL >= 2 :  {29654693: 5, 29678332: 2}\n",
      "230106\n",
      "INDEL >= 2 :  {29603981: 4, 29639197: 3, 29639176: 2, 29661364: 2, 29682995: 3, 29683062: 3}\n",
      "230127\n",
      "INDEL >= 2 :  {29654671: 2, 29661417: 2, 29665067: 3, 29668289: 2, 29671853: 2, 29671960: 2}\n",
      "230217\n",
      "INDEL >= 2 :  {29639310: 2, 29642053: 2}\n",
      "230303\n",
      "INDEL >= 2 :  {29664999: 2, 29674746: 2, 29678154: 2, 29681461: 2, 29681596: 2}\n",
      "230323\n",
      "INDEL >= 2 :  {29661336: 75, 29668454: 4, 29668372: 3, 29681451: 2, 29682985: 3, 29683079: 6, 29683060: 6, 29683075: 3}\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import pysam\n",
    "import gzip\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "\n",
    "sample_name = [line.rstrip('\\n') for line in open('/data/project/Meningioma/script/00.Targetted/sample_name.txt', 'r')]\n",
    "\n",
    "start_end_list = []\n",
    "f = open(\"/home/goldpm1/resources/NF2.exome.splicing.bed\",'r')\n",
    "for fline in f:\n",
    "    ftab = fline.rstrip().split('\\t')\n",
    "    start = ftab[1]\n",
    "    end = ftab[2]\n",
    "    element = start + '_' + end\n",
    "    start_end_list.append(element)\n",
    "f.close()\n",
    "\n",
    "reference = pysam.FastaFile('/home/goldpm1/reference/genome.fa')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for SAMPLE in sample_name:\n",
    "    print (SAMPLE)\n",
    "\n",
    "    bam_dir = '/data/project/Meningioma/00.Targetted/02.Align/hg38/Dura/05.Final_bam'\n",
    "    \n",
    "    samfile = pysam.AlignmentFile(bam_dir + \"/\" + SAMPLE + \"_Dura.bam\", \"rb\")\n",
    "    dict = {}\n",
    "    dict_mismatch = {}\n",
    "\n",
    "    # INDEL\n",
    "    for start_end in start_end_list:\n",
    "        start_pos = int(start_end.rstrip().split('_')[0])\n",
    "        end_pos = int(start_end.rstrip().split('_')[1])\n",
    "\n",
    "        iter = samfile.fetch(contig = 'chr22', start = start_pos - 1, end = end_pos + 1)\n",
    "        for i, read in enumerate(iter):\n",
    "            if read.is_paired:\n",
    "                read_info = str(read).rstrip().split('\\t')\n",
    "                        \n",
    "                soft_al = read.query_alignment_start\n",
    "                if read.cigarstring != None:  # 왜 None이 있지?\n",
    "                    if (\"D\" in read.cigarstring)   |  (\"I\" in read.cigarstring) :\n",
    "                        groups = re.findall(r\"(\\d+|\\D+)\", read.cigarstring)\n",
    "                    else:\n",
    "                        continue\n",
    "                    \n",
    "                    sum = 0 \n",
    "                    if \"D\" in groups:\n",
    "                        end = groups.index(\"D\")\n",
    "                    elif \"I\" in groups:\n",
    "                        end = groups.index(\"I\")\n",
    "\n",
    "                    for i in range (0, end - 1 ):\n",
    "                        if groups[i].isdigit():  # 숫자면 더해준다\n",
    "                            sum = sum + int(groups[i])\n",
    "\n",
    "                    if int (read.pos) + sum + 1  in dict.keys():\n",
    "                        dict [ int (read.pos) + sum  + 1] += 1    # 있으면 그 key에 1씩 더해줌\n",
    "                    else:\n",
    "                        dict [ int (read.pos) + sum  + 1] = 1       # 없으면 1\n",
    "\n",
    "\n",
    "    filtered_dict = {key: value for key, value in dict.items() if value  >= 2}\n",
    "    print ( \"INDEL >= 2 : \", filtered_dict  )\n",
    "\n",
    "    \n",
    "    # # 모든 read를 돌면서 multiallelic mismatch (variant)가 있는지 본다\n",
    "    # for start_end in start_end_list:\n",
    "    #     start_pos = int(start_end.rstrip().split('_')[0])\n",
    "    #     end_pos = int(start_end.rstrip().split('_')[1])\n",
    "\n",
    "    #     iter = samfile.fetch(contig = 'chr22', start = start_pos - 1, end = end_pos + 1)\n",
    "    #     for i, read in enumerate(iter):  # 해당 splicing region의 read를 모두 소집\n",
    "    #         reference_sequence = reference.fetch(read.reference_name, read.reference_start, read.reference_end)  # 그 read에 맞는 reference sequence\n",
    "\n",
    "    #         i = 0\n",
    "    #         for read_base, ref_base in zip(read.query_sequence, reference_sequence):\n",
    "    #             if read_base != ref_base:\n",
    "    #                 if read.reference_start + i in  dict_mismatch.keys():\n",
    "    #                     dict_mismatch[read.reference_start + i].add ( read_base )  # set에 추가\n",
    "    #                 else:\n",
    "    #                     dict_mismatch[read.reference_start + i]  =  set(read_base)   # 1개짜리 set를 만들어서 넣어준다\n",
    "    #             i = i + 1\n",
    "\n",
    "    # if any(len(value) > 2 for value in dict_mismatch.values()) == True:\n",
    "    #     dict_mismatch = {key:value for key, value in dict_mismatch.items() if len(value) >= 2 }\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'A': 1144, 'C': 0, 'G': 0, 'T': 0}\n"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "\n",
    "# Open the BAM file\n",
    "bam_file = pysam.AlignmentFile( '/data/project/Meningioma/00.Targetted/02.Align/hg38/Dura/05.Final_bam/221021_Dura.bam', 'r')\n",
    "\n",
    "# Define the target position\n",
    "chromosome = 'chr22'\n",
    "position = 29639209\n",
    "\n",
    "# Count the base distribution\n",
    "base_counts = {'A': 0, 'C': 0, 'G': 0, 'T': 0}\n",
    "\n",
    "for pileup_column in bam_file.pileup(chromosome, position - 1, position):\n",
    "    if pileup_column.pos == position - 1:\n",
    "        #print ( len( pileup_column.pileups )  )\n",
    "        for pileup_read in pileup_column.pileups:\n",
    "            if pileup_read.query_position == None:\n",
    "                continue\n",
    "            base = pileup_read.alignment.query_sequence[pileup_read.query_position]\n",
    "            base_counts [base] += 1\n",
    "\n",
    "print (base_counts)\n",
    "bam_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NF2 splicing region bed : /home/goldpm1/resources/NF2.exome.splicing.bed'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr22\t29603551\t29603559\tsplicing_region\n",
      "chr22\t29603583\t29603591\tsplicing_region\n",
      "chr22\t29603598\t29603606\tsplicing_region\n",
      "chr22\t29603612\t29603620\tsplicing_region\n",
      "chr22\t29603994\t29604002\tsplicing_region\n",
      "chr22\t29604108\t29604116\tsplicing_region\n",
      "chr22\t29628653\t29628661\tsplicing_region\n",
      "chr22\t29628769\t29628777\tsplicing_region\n",
      "chr22\t29636746\t29636754\tsplicing_region\n",
      "chr22\t29636872\t29636880\tsplicing_region\n",
      "chr22\t29639085\t29639093\tsplicing_region\n",
      "chr22\t29639208\t29639216\tsplicing_region\n",
      "chr22\t29642197\t29642205\tsplicing_region\n",
      "chr22\t29642281\t29642289\tsplicing_region\n",
      "chr22\t29654652\t29654660\tsplicing_region\n",
      "chr22\t29654721\t29654729\tsplicing_region\n",
      "chr22\t29654970\t29654978\tsplicing_region\n",
      "chr22\t29655076\t29655084\tsplicing_region\n",
      "chr22\t29658184\t29658192\tsplicing_region\n",
      "chr22\t29658260\t29658268\tsplicing_region\n",
      "chr22\t29661200\t29661208\tsplicing_region\n",
      "chr22\t29661335\t29661343\tsplicing_region\n",
      "chr22\t29664985\t29664993\tsplicing_region\n",
      "chr22\t29665060\t29665068\tsplicing_region\n",
      "chr22\t29668328\t29668336\tsplicing_region\n",
      "chr22\t29668442\t29668450\tsplicing_region\n",
      "chr22\t29671821\t29671829\tsplicing_region\n",
      "chr22\t29671944\t29671952\tsplicing_region\n",
      "chr22\t29673264\t29673272\tsplicing_region\n",
      "chr22\t29673482\t29673490\tsplicing_region\n",
      "chr22\t29674831\t29674839\tsplicing_region\n",
      "chr22\t29674937\t29674945\tsplicing_region\n",
      "chr22\t29678191\t29678199\tsplicing_region\n",
      "chr22\t29678319\t29678327\tsplicing_region\n",
      "chr22\t29681434\t29681442\tsplicing_region\n",
      "chr22\t29681597\t29681605\tsplicing_region\n",
      "chr22\t29683015\t29683023\tsplicing_region\n",
      "chr22\t29683060\t29683068\tsplicing_region\n",
      "chr22\t29683075\t29683083\tsplicing_region\n",
      "chr22\t29683696\t29683704\tsplicing_region\n",
      "chr22\t29683911\t29683919\tsplicing_region\n",
      "chr22\t29684158\t29684166\tsplicing_region\n",
      "chr22\t29694747\t29694755\tsplicing_region\n",
      "chr22\t29698594\t29698602\tsplicing_region\n"
     ]
    }
   ],
   "source": [
    "start_end_list = []\n",
    "input_bed = open('/home/goldpm1/resources/NF2.exome.proteincoding.bed','r')\n",
    "for fline in input_bed:\n",
    "    ftab = fline.rstrip().split('\\t')\n",
    "    start = ftab[1]\n",
    "    end = ftab[2]\n",
    "    element_1 = str ( ( int(start) - 4) ) + '_' + str ( ( int(start) + 4) )\n",
    "    element_2 = str ( ( int(end) - 4) ) + '_' + str ( ( int(end) + 4) )\n",
    "    start_end_list.append(element_1)\n",
    "    start_end_list.append(element_2)\n",
    "input_bed.close()\n",
    "\n",
    "start_end_list = sorted( list(set(start_end_list)))\n",
    "\n",
    "\n",
    "output_file = open('/home/goldpm1/resources/NF2.exome.splicing.bed','w')\n",
    "for line in start_end_list:\n",
    "    print (\"chr22\\t{}\\t{}\\tsplicing_region\".format (line.split(\"_\")[0], line.split(\"_\")[1]), file = output_file)\n",
    "    #print (\"chr22\\t{}\\t{}\\tsplicing_region\".format (line.split(\"_\")[0], line.split(\"_\")[1]))\n",
    "output_file.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cnvpytor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
