{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 24\n",
      "1 15\n",
      "2 30\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy\n",
    "\n",
    "df = pd.read_csv ( \"/data/project/Meningioma/31.Clonality/02.pyclonevi/230419/230419.decomposed-facetcnv.tsv\", sep = \"\\t\" )\n",
    "df_count = np.unique ( df [\"cluster_id\"], return_counts = True )\n",
    "\n",
    "for i in range ( len(df_count[0]) ) :\n",
    "    print (df_count[0][i], int(df_count[1][i] / len (df[\"sample_id\"].unique() )) )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 48\n",
      "1 30\n",
      "2 60\n"
     ]
    }
   ],
   "source": [
    "for i in range ( len(df_count[0]) ) :\n",
    "    print (df_count[0][i], df_count[1][i] / len (df[\"sample_id\"].unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_id</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>ref_counts</th>\n",
       "      <th>alt_counts</th>\n",
       "      <th>normal_cn</th>\n",
       "      <th>major_cn</th>\n",
       "      <th>minor_cn</th>\n",
       "      <th>tumour_content</th>\n",
       "      <th>types</th>\n",
       "      <th>gene</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>cluster_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_116124115</td>\n",
       "      <td>230419_Dura</td>\n",
       "      <td>378</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>MAB21L3</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_116124115</td>\n",
       "      <td>230419_Tumor</td>\n",
       "      <td>378</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>MAB21L3</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_204229027</td>\n",
       "      <td>230419_Dura</td>\n",
       "      <td>233</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>PLEKHA6</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_204229027</td>\n",
       "      <td>230419_Tumor</td>\n",
       "      <td>233</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>PLEKHA6</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_74480652</td>\n",
       "      <td>230419_Dura</td>\n",
       "      <td>174</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>LRRC53</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>chr19_45616430</td>\n",
       "      <td>230419_Tumor</td>\n",
       "      <td>157</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>EML2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>chr20_37066790</td>\n",
       "      <td>230419_Dura</td>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>RBL1</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>chr20_37066790</td>\n",
       "      <td>230419_Tumor</td>\n",
       "      <td>104</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>RBL1</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>chr22_43171854</td>\n",
       "      <td>230419_Dura</td>\n",
       "      <td>64</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>TTLL12</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>chr22_43171854</td>\n",
       "      <td>230419_Tumor</td>\n",
       "      <td>64</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Mutect2</td>\n",
       "      <td>TTLL12</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        mutation_id     sample_id  ref_counts  alt_counts  normal_cn  \\\n",
       "0    chr1_116124115   230419_Dura         378           5          2   \n",
       "1    chr1_116124115  230419_Tumor         378           0          2   \n",
       "2    chr1_204229027   230419_Dura         233           8          2   \n",
       "3    chr1_204229027  230419_Tumor         233           0          2   \n",
       "4     chr1_74480652   230419_Dura         174           5          2   \n",
       "..              ...           ...         ...         ...        ...   \n",
       "133  chr19_45616430  230419_Tumor         157           5          2   \n",
       "134  chr20_37066790   230419_Dura         104           0          2   \n",
       "135  chr20_37066790  230419_Tumor         104           6          2   \n",
       "136  chr22_43171854   230419_Dura          64           0          2   \n",
       "137  chr22_43171854  230419_Tumor          64           4          2   \n",
       "\n",
       "     major_cn  minor_cn  tumour_content    types     gene  \\\n",
       "0           1         1            1.00  Mutect2  MAB21L3   \n",
       "1           1         1            0.29  Mutect2  MAB21L3   \n",
       "2           1         1            1.00  Mutect2  PLEKHA6   \n",
       "3           1         1            0.29  Mutect2  PLEKHA6   \n",
       "4           1         1            1.00  Mutect2   LRRC53   \n",
       "..        ...       ...             ...      ...      ...   \n",
       "133         1         0            0.29  Mutect2     EML2   \n",
       "134         1         0            1.00  Mutect2     RBL1   \n",
       "135         1         0            0.29  Mutect2     RBL1   \n",
       "136         1         0            1.00  Mutect2   TTLL12   \n",
       "137         1         0            0.29  Mutect2   TTLL12   \n",
       "\n",
       "    variant_classification  cluster_id  \n",
       "0         missense_variant           0  \n",
       "1         missense_variant           0  \n",
       "2       synonymous_variant           0  \n",
       "3       synonymous_variant           0  \n",
       "4       synonymous_variant           0  \n",
       "..                     ...         ...  \n",
       "133         intron_variant           2  \n",
       "134     synonymous_variant           2  \n",
       "135     synonymous_variant           2  \n",
       "136       missense_variant           2  \n",
       "137       missense_variant           2  \n",
       "\n",
       "[138 rows x 12 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chr1', '150487535', '.', 'A', 'G']\n",
      "['chr3', '50194058', '.', 'C', 'A']\n",
      "['chr3', '52322739', '.', 'C', 'T']\n",
      "['chr5', '103108723', '.', 'A', 'G']\n",
      "['chr6', '32554712', '.', 'C', 'T']\n",
      "['chr12', '57730658', '.', 'C', 'T']\n",
      "['chr12', '106860433', '.', 'T', 'G']\n",
      "['chr12', '131165351', '.', 'G', 'A']\n",
      "['chr13', '28438321', '.', 'A', 'G']\n",
      "['chr13', '113623173', '.', 'G', 'A']\n",
      "['chr15', '43876007', '.', 'C', 'T']\n",
      "['chr15', '48332213', '.', 'T', 'G']\n",
      "['chr15', '78484821', '.', 'T', 'TA']\n",
      "['chr16', '2235405', '.', 'G', 'A']\n",
      "['chr17', '7254728', '.', 'T', 'C']\n",
      "['chr17', '76478828', '.', 'T', 'C']\n",
      "['chr19', '55341955', '.', 'A', 'G']\n",
      "['chr21', '25741676', '.', 'GT', 'G']\n"
     ]
    }
   ],
   "source": [
    "import vcf, pysam, pybedtools\n",
    "\n",
    "tbx = pysam.TabixFile( \"/home/goldpm1/Meningioma/04.mutect/03.vep/220930_Dura.MT2.FMC.HF.RMBLACK.vep.vcf.gz\" )\n",
    "\n",
    "a = pybedtools.BedTool('/home/goldpm1/Meningioma/04.mutect/03.vep/220930_multiple.MT2.FMC.HF.RMBLACK.vep.vcf')\n",
    "b = pybedtools.BedTool(\"/home/goldpm1/Meningioma/04.mutect/03.vep/220930_Dura.MT2.FMC.HF.RMBLACK.vep.vcf\")\n",
    "\n",
    "ab = a.intersect(b, v = True) \n",
    "for interval in ab:\n",
    "    print ( interval[0:5], sep = \"\\t\" )\n",
    "\n",
    "# if not tbx.fetch(\"chr12\", 57730658, 57730659) :\n",
    "#     print (\"hello\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequenza → PycloneVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1_2787605\t220930_Dura\t190\t5\t2\t1\t1\t0.3\tMutect2\tTTC34\tmissense_variant\n",
      "chr1_42154829\t220930_Dura\t186\t5\t2\t1\t1\t0.3\tMutect2\tGUCA2B\tsynonymous_variant\n",
      "chr1_44802872\t220930_Dura\t98\t14\t2\t1\t1\t0.3\tMutect2\tTCTEX1D4\tdownstream_gene_variant\n",
      "chr1_243221776\t220930_Dura\t492\t48\t2\t1\t1\t0.3\tMutect2\tCEP170\tmissense_variant\n",
      "chr2_3644016\t220930_Dura\t830\t63\t2\t1\t1\t0.3\tMutect2\tCOLEC11\tsynonymous_variant\n",
      "chr2_44706287\t220930_Dura\t278\t16\t2\t1\t1\t0.3\tMutect2\tCAMKMT\tmissense_variant&splice_region_variant\n",
      "chr2_178554137\t220930_Dura\t603\t19\t2\t1\t1\t0.3\tMutect2\tTTN\tsynonymous_variant\n",
      "chr2_185808060\t220930_Dura\t113\t5\t2\t1\t1\t0.3\tMutect2\tFSIP2\tintron_variant&NMD_transcript_variant\n",
      "chr2_222909077\t220930_Dura\t85\t5\t2\t1\t1\t0.3\tMutect2\tACSL3\tmissense_variant\n",
      "chr2_241240049\t220930_Dura\t187\t5\t2\t1\t1\t0.3\tMutect2\tHDLBP\tmissense_variant\n",
      "chr3_11817144\t220930_Dura\t363\t20\t2\t1\t1\t0.3\tMutect2\tTAMM41\tintron_variant\n",
      "chr3_49721007\t220930_Dura\t397\t25\t2\t1\t1\t0.3\tMutect2\tGMPPB\t3_prime_UTR_variant\n",
      "chr5_71656712\t220930_Dura\t172\t8\t2\t1\t1\t0.3\tMutect2\tMCCC2\tintron_variant\n",
      "chr6_87255842\t220930_Dura\t257\t15\t2\t1\t1\t0.3\tMutect2\tZNF292\tmissense_variant\n",
      "chr7_106026646\t220930_Dura\t419\t24\t2\t1\t1\t0.3\tMutect2\tCDHR3\tintron_variant\n",
      "chr7_122386320\t220930_Dura\t39\t4\t2\t1\t1\t0.3\tMutect2\tCADPS2\tintron_variant\n",
      "chr8_10622949\t220930_Dura\t168\t11\t2\t1\t1\t0.3\tMutect2\tRP1L1\tnon_coding_transcript_exon_variant\n",
      "chr8_81801465\t220930_Dura\t121\t5\t2\t1\t1\t0.3\tMutect2\tSNX16\t3_prime_UTR_variant\n",
      "chr8_85126674\t220930_Dura\t323\t21\t2\t1\t1\t0.3\tMutect2\tLRRCC1\tintron_variant\n",
      "chr9_127673284\t220930_Dura\t129\t5\t2\t1\t1\t0.3\tMutect2\tSTXBP1\tintron_variant\n",
      "chr10_34360246\t220930_Dura\t262\t11\t2\t1\t1\t0.3\tMutect2\tPARD3\tmissense_variant\n",
      "chr15_40252764\t220930_Dura\t193\t12\t2\t1\t1\t0.3\tMutect2\tPAK6\tintron_variant\n",
      "chr15_43524923\t220930_Dura\t308\t22\t2\t1\t1\t0.3\tMutect2\tMAP1A\tsynonymous_variant\n",
      "chr16_3679864\t220930_Dura\t282\t24\t2\t1\t0\t0.3\tMutect2\tTRAP1\tintron_variant\n",
      "chr17_38361755\t220930_Dura\t255\t27\t2\t1\t1\t0.3\tMutect2\tSOCS7\tmissense_variant\n",
      "chr18_54884282\t220930_Dura\t309\t21\t2\t1\t1\t0.3\tMutect2\tRAB27B\tintron_variant\n",
      "chr18_74559301\t220930_Dura\t147\t5\t2\t1\t1\t0.3\tMutect2\tCNDP1\tintron_variant\n",
      "chr20_13849480\t220930_Dura\t221\t7\t2\t1\t1\t0.3\tMutect2\tSEL1L2\t3_prime_UTR_variant\n",
      "chr22_29604113\t220930_Dura\t298\t21\t2\t1\t0\t0.3\tMutect2\tNF2\tsplice_donor_variant\n",
      "chr1_160041721\t220930_Dura\t182\t150\t2\t1\t1\t0.3\tHC\tKCNJ10\tintron_variant\n",
      "chr1_161042970\t220930_Dura\t57\t68\t2\t1\t1\t0.3\tHC\tAL591806.3\tupstream_gene_variant\n",
      "chr1_233650040\t220930_Dura\t124\t109\t2\t1\t1\t0.3\tHC\tKCNK1\tintron_variant&non_coding_transcript_variant\n",
      "chr2_218278604\t220930_Dura\t184\t148\t2\t1\t1\t0.3\tHC\tTMBIM1\tintron_variant\n",
      "chr2_233845860\t220930_Dura\t190\t146\t2\t1\t1\t0.3\tHC\tHJURP\tnon_coding_transcript_exon_variant\n",
      "chr2_240465082\t220930_Dura\t122\t138\t2\t1\t1\t0.3\tHC\tGPC1\tsynonymous_variant\n",
      "chr3_42739784\t220930_Dura\t68\t95\t2\t1\t1\t0.3\tHC\tCCDC13\tsynonymous_variant\n",
      "chr3_50186160\t220930_Dura\t76\t91\t2\t1\t1\t0.3\tHC\tSEMA3F\tintron_variant\n",
      "chr3_195963380\t220930_Dura\t335\t145\t2\t1\t1\t0.3\tHC\tSDHAP1\tnon_coding_transcript_exon_variant\n",
      "chr4_1983530\t220930_Dura\t235\t186\t2\t1\t1\t0.3\tHC\tNSD2\tdownstream_gene_variant\n",
      "chr5_7867917\t220930_Dura\t225\t156\t2\t1\t1\t0.3\tHC\tMTRR\tupstream_gene_variant\n",
      "chr5_69392781\t220930_Dura\t61\t72\t2\t1\t1\t0.3\tHC\tRAD17\tintron_variant\n",
      "chr5_96788593\t220930_Dura\t116\t64\t2\t1\t1\t0.3\tHC\tERAP1\tsynonymous_variant\n",
      "chr6_32581859\t220930_Dura\t200\t53\t2\t1\t1\t0.3\tHC\tHLA-DRB1\tintron_variant\n",
      "chr6_104852387\t220930_Dura\t72\t50\t2\t1\t1\t0.3\tHC\tHACE1\tintron_variant\n",
      "chr7_100959268\t220930_Dura\t431\t135\t2\t1\t1\t0.3\tHC\tMUC3A\tmissense_variant\n",
      "chr7_149806755\t220930_Dura\t180\t170\t2\t1\t1\t0.3\tHC\tSSPOP\tnon_coding_transcript_exon_variant\n",
      "chr8_143918193\t220930_Dura\t117\t99\t2\t1\t1\t0.3\tHC\tPLEC\tsynonymous_variant\n",
      "chr9_34834261\t220930_Dura\t132\t23\t2\t1\t1\t0.3\tHC\tFAM205BP\tdownstream_gene_variant\n",
      "chr10_70775309\t220930_Dura\t179\t162\t2\t1\t1\t0.3\tHC\tTBATA\tintron_variant\n",
      "chr10_113650475\t220930_Dura\t100\t104\t2\t1\t1\t0.3\tHC\tNRAP\tmissense_variant\n",
      "chr11_618172\t220930_Dura\t134\t161\t2\t1\t1\t0.3\tHC\tIRF7\tupstream_gene_variant\n",
      "chr11_1082445\t220930_Dura\t137\t146\t2\t1\t1\t0.3\tHC\tMUC2\tintron_variant&non_coding_transcript_variant\n",
      "chr11_26563078\t220930_Dura\t95\t83\t2\t1\t1\t0.3\tHC\tANO3\tintron_variant\n",
      "chr11_56701223\t220930_Dura\t501\t361\t2\t1\t1\t0.3\tHC\tOR9G1\tmissense_variant\n",
      "chr11_62721795\t220930_Dura\t167\t144\t2\t1\t1\t0.3\tHC\tHNRNPUL2\tintron_variant\n",
      "chr11_70161378\t220930_Dura\t134\t97\t2\t1\t1\t0.3\tHC\tANO1\tintron_variant\n",
      "chr12_31451994\t220930_Dura\t193\t163\t2\t1\t1\t0.3\tHC\tDENND5B\tsynonymous_variant\n",
      "chr14_22956973\t220930_Dura\t107\t139\t2\t1\t0\t0.3\tHC\tHAUS4\t5_prime_UTR_variant\n",
      "chr14_78243471\t220930_Dura\t209\t228\t2\t1\t0\t0.3\tHC\tNRXN3\tsynonymous_variant\n",
      "chr14_106062599\t220930_Dura\t206\t254\t2\t1\t0\t0.3\tHC\tIGHV3-7\tsynonymous_variant\n",
      "chr15_51210647\t220930_Dura\t175\t168\t2\t1\t1\t0.3\tHC\tCYP19A1\t3_prime_UTR_variant\n",
      "chr15_53702296\t220930_Dura\t133\t95\t2\t1\t1\t0.3\tHC\tWDR72\tsynonymous_variant\n",
      "chr15_64500333\t220930_Dura\t131\t123\t2\t1\t1\t0.3\tHC\tZNF609\tintron_variant\n",
      "chr15_90084384\t220930_Dura\t70\t82\t2\t1\t1\t0.3\tHC\tZNF710\tdownstream_gene_variant\n",
      "chr15_100569472\t220930_Dura\t253\t230\t2\t1\t1\t0.3\tHC\tLINS1\tmissense_variant\n",
      "chr16_1342876\t220930_Dura\t104\t81\t2\t1\t0\t0.3\tHC\tBAIAP3\tintron_variant\n",
      "chr16_21983090\t220930_Dura\t114\t88\t2\t1\t0\t0.3\tHC\tUQCRC2\tsplice_region_variant&synonymous_variant\n",
      "chr16_22138588\t220930_Dura\t103\t122\t2\t1\t0\t0.3\tHC\tVWA3A\tintron_variant&NMD_transcript_variant\n",
      "chr16_22250579\t220930_Dura\t92\t80\t2\t1\t0\t0.3\tHC\tEEF2K\tintron_variant\n",
      "chr17_40032802\t220930_Dura\t108\t94\t2\t1\t1\t0.3\tHC\tMED24\tintron_variant\n",
      "chr18_674320\t220930_Dura\t76\t59\t2\t1\t1\t0.3\tHC\tTYMS\tdownstream_gene_variant\n",
      "chr18_693977\t220930_Dura\t106\t74\t2\t1\t1\t0.3\tHC\tENOSF1\tintron_variant\n",
      "chr19_579450\t220930_Dura\t89\t91\t2\t1\t0\t0.3\tHC\tBSG\tintron_variant\n",
      "chr19_20039366\t220930_Dura\t124\t89\t2\t1\t0\t0.3\tHC\tZNF682\tupstream_gene_variant\n",
      "chr19_54806001\t220930_Dura\t145\t192\t2\t1\t0\t0.3\tHC\tKIR2DL4\tmissense_variant\n",
      "chr20_1636246\t220930_Dura\t329\t302\t2\t1\t1\t0.3\tHC\tSIRPG\tsynonymous_variant\n",
      "chr20_32189533\t220930_Dura\t117\t100\t2\t1\t1\t0.3\tHC\tPLAGL2\tdownstream_gene_variant\n",
      "chr20_45886700\t220930_Dura\t89\t65\t2\t1\t1\t0.3\tHC\tCTSA\tupstream_gene_variant\n",
      "chr21_46534147\t220930_Dura\t206\t175\t2\t1\t1\t0.3\tHC\tDIP2A\tintron_variant\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import pybedtools\n",
    "import pysam\n",
    "import argparse\n",
    "import random\n",
    "\n",
    "SEQUENZA_MUTATION_PATH=\"/home/goldpm1/Meningioma/11.cnv/2.sequenza/220930_Dura_mutations.txt\"\n",
    "SEQUENZA_SEGMENT_PATH=\"/home/goldpm1/Meningioma/11.cnv/2.sequenza/220930_Dura_segments.txt\"\n",
    "SEQUENZA_PLOIDY_PATH=\"/home/goldpm1/Meningioma/11.cnv/2.sequenza/220930_Dura_confints_CP.txt\"\n",
    "SEQUENZA_TO_PYCLONEVI_MATRIX_PATH=\"/home/goldpm1/Meningioma/31.Clonality/01.make_matrix/220930/220930.sequenza_to_pyclonevi.tsv\"\n",
    "MUTECT_OUTPUT_PATH=\"/home/goldpm1/Meningioma/04.mutect/02.PASS/220930_Dura.MT2.FMC.HF.vcf\"\n",
    "MUTECT_OUTPUT_PATH=\"/home/goldpm1/Meningioma/04.mutect/03.vep/220930_Dura.MT2.FMC.HF.RMBLACK.vep.vcf\"\n",
    "GVCF_OUTPUT_PATH=\"/home/goldpm1/Meningioma/05.gvcf/02.remove_nonref/220930/220930_Dura.g.vcf.gz\"\n",
    "HC_OUTPUT_PATH=\"/home/goldpm1/Meningioma/06.hc/04.vep/220930/Dura/220930_Dura.DP100.vep.vcf\"\n",
    "HC_BLOOD_RANDOM_PICK_PATH=\"/home/goldpm1/Meningioma/31.Clonality/01.make_matrix/220930/220930.HC.random_pick_50.bed\"\n",
    "TISSUE=\"Dura\"\n",
    "Sample_ID=\"220930\"\n",
    "\n",
    "tbx = pysam.TabixFile (GVCF_OUTPUT_PATH)\n",
    "\n",
    "SAMPLE_ID = Sample_ID + \"_\" + TISSUE\n",
    "\n",
    "ploidy_df = pd.read_csv (SEQUENZA_PLOIDY_PATH, sep = \"\\t\")\n",
    "TUMOR_PURITY = ploidy_df.iloc[0][\"cellularity\"]\n",
    "\n",
    "mutation_df = pd.read_csv (SEQUENZA_MUTATION_PATH, sep = \"\\t\")  \n",
    "segment_df = pd.read_csv (SEQUENZA_SEGMENT_PATH, sep = \"\\t\")\n",
    "\n",
    "mutation_df[\"end\"] = mutation_df[\"position\"]\n",
    "mutation_df[\"start\"] = mutation_df[\"position\"] - 1\n",
    "mutation_df = mutation_df.reindex (columns = [\"chromosome\", \"start\", \"end\", \"GC.percent\", \"good.reads\", \"adjusted.ratio\", \"F\", \"mutation\", \"CNn\", \"CNt\", \"Mt\"] )\n",
    "mutation_df.head()\n",
    "\n",
    "\n",
    "######## Mutect call을 골라줌 ###################\n",
    "mutect_pybed_object = pybedtools.BedTool(MUTECT_OUTPUT_PATH)\n",
    "segment_pybed_object =  pybedtools.BedTool.from_dataframe(segment_df)\n",
    "\n",
    "ab = segment_pybed_object.intersect(mutect_pybed_object, wb = True)\n",
    "\n",
    "for index, contents in enumerate( ab[0] ):\n",
    "    if \"GT:\" in contents:\n",
    "        parsing_sample_index = index \n",
    "        break\n",
    "\n",
    "output_file = open(SEQUENZA_TO_PYCLONEVI_MATRIX_PATH, \"w\")\n",
    "\n",
    "for interval in ab:\n",
    "    CHR, POS = interval[0], interval[2]\n",
    "    line_dict = {}\n",
    "\n",
    "    line_dict [\"mutation_id\"] = str(CHR) + \"_\"  + str(POS)\n",
    "    line_dict [\"sample_id\"] = SAMPLE_ID\n",
    "    line_dict [\"ref_counts\"] = str( interval[parsing_sample_index + 2].split(\":\")[1].split(\",\")[0])\n",
    "    line_dict [\"alt_counts\"] = str( interval[parsing_sample_index + 2].split(\":\")[1].split(\",\")[1])\n",
    "    line_dict[\"normal_cn\"] = str ( 2 )\n",
    "    line_dict[\"major_cn\"] = str ( interval[10] )\n",
    "    line_dict[\"minor_cn\"] = str ( interval[11] )\n",
    "    line_dict[\"tumour_content\"] = str( TUMOR_PURITY  )\n",
    "    line_dict[\"type\"] = \"Mutect2\"\n",
    "    line_dict[\"gene\"] = str( interval [parsing_sample_index - 1] ).split(\";\")[-1].split(\"|\")[3]\n",
    "    line_dict[\"Variant_Classification\"] = str( interval [parsing_sample_index - 1] ).split(\";\")[-1].split(\"|\")[1]\n",
    "    \n",
    "    print ( '\\t'.join ( list( line_dict.values()) ) )\n",
    "    print ( '\\t'.join ( list( line_dict.values()) ), file = output_file )\n",
    "        \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "   \n",
    "    \n",
    "######## HC call을 골라줌 ###################\n",
    "hc_blood_pybed_object = pybedtools.BedTool ( HC_BLOOD_RANDOM_PICK_PATH )  \n",
    "hc_pybed_object = pybedtools.BedTool(HC_OUTPUT_PATH)  #VCF\n",
    "\n",
    "hc_selected_pybed_object = hc_blood_pybed_object.intersect(hc_pybed_object, wa = True, wb = True)       # Dura/Tumor에서 발견된 HC call 중에서 선택된 50개만 고름 (concordance가 95%이기 때문에 47개정도 예상)\n",
    "\n",
    "\n",
    "ab = hc_selected_pybed_object.intersect(segment_pybed_object, wa = True, wb = True)\n",
    "#print (ab [0])\n",
    "\n",
    "for index, contents in enumerate( ab[0] ):\n",
    "    if \"GT:\" in contents:\n",
    "        parsing_sample_index = index \n",
    "        break\n",
    "\n",
    "\n",
    "for interval in ab:\n",
    "    CHR, POS = interval[0], int (interval[2]) - 1\n",
    "    line_dict = {}\n",
    "\n",
    "    line_dict [\"mutation_id\"] = str(CHR) + \"_\"  + str(POS)\n",
    "    line_dict [\"sample_id\"] = SAMPLE_ID\n",
    "    line_dict [\"ref_counts\"] = str( interval [parsing_sample_index + 1] ).split(\":\")[1].split(\",\")[0]\n",
    "    line_dict [\"alt_counts\"] = str( interval [parsing_sample_index + 1] ).split(\":\")[1].split(\",\")[1]\n",
    "    line_dict[\"normal_cn\"] = str ( 2 )\n",
    "    line_dict[\"major_cn\"] = str ( interval[ -3 ] )\n",
    "    line_dict[\"minor_cn\"] = str ( interval[ -2 ] )\n",
    "    line_dict[\"tumour_content\"] = str( TUMOR_PURITY  )\n",
    "    line_dict[\"type\"] = \"HC\"\n",
    "    line_dict[\"gene\"] = str( interval [parsing_sample_index - 1] ).split(\";\")[-1].split(\"|\")[3]\n",
    "    line_dict[\"Variant_Classification\"] = str( interval [parsing_sample_index - 1] ).split(\";\")[-1].split(\"|\")[1]\n",
    "    #line_dict[\"gene\"] = str( interval [parsing_sample_index - 1] ).split(\":\")[-1].split(\"|\")[3]\n",
    "    \n",
    "    print ( '\\t'.join ( list( line_dict.values()) ) )\n",
    "    print ( '\\t'.join ( list( line_dict.values()) ), file = output_file )\n",
    "   \n",
    "\n",
    "output_file.close()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FacetCNV → PycloneVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/goldpm1/Meningioma/11.cnv/5.facetcnv/220930/Dura/220930.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gzip\n",
    "import os, pybedtools\n",
    "\n",
    "FACETCNV_OUTPUT_PATH=\"/home/goldpm1/Meningioma/11.cnv/5.facetcnv/220930/Dura/220930.vcf.gz\"\n",
    "FACETCNV_TO_PYCLONEVI_MATRIX_PATH=\"/home/goldpm1/Meningioma/31.Clonality/01.make_matrix/220930/220930.facetcnv_to_pyclonevi.tsv\"\n",
    "MUTECT_OUTPUT_PATH=\"/home/goldpm1/Meningioma/04.mutect/03.vep/220930_Dura.MT2.FMC.HF.RMBLACK.vep.vcf\"\n",
    "HC_OUTPUT_PATH=\"/home/goldpm1/Meningioma/06.hc/01.call/220930/Dura/220930_Dura.vcf\"\n",
    "HC_BLOOD_RANDOM_PICK_PATH=\"/home/goldpm1/Meningioma/31.Clonality/01.make_matrix/220930/220930.HC.random_pick_50.bed\"\n",
    "TISSUE=\"Dura\"\n",
    "Sample_ID=\"220930\"\n",
    "\n",
    "SAMPLE_ID = Sample_ID + \"_\" + TISSUE\n",
    "\n",
    "\n",
    "\n",
    "if \"gz\" in FACETCNV_OUTPUT_PATH:\n",
    "    input_file = gzip.open (FACETCNV_OUTPUT_PATH, \"r\")\n",
    "else:\n",
    "    input_file = open (FACETCNV_OUTPUT_PATH, \"r\")\n",
    "sample_name = []\n",
    "\n",
    "\n",
    "def parsing (line):\n",
    "    CHR, POS, REF, ALT = line[0], int(line[1]), line[3], line[4]\n",
    "    # 7. info\n",
    "    info_list = line[7].split(';')\n",
    "    info_dict = {}\n",
    "    for i in range( len(info_list) ):\n",
    "        if \"=\" in info_list[i]:\n",
    "            info_dict [ info_list[i].split('=')[0] ] = info_list[i].split('=')[1]\n",
    "\n",
    "    return CHR, POS, REF, ALT, info_list, info_dict\n",
    "\n",
    "\n",
    "\n",
    "matrix = [] \n",
    "colnames = ['CHR','START','END', 'MAJOR_CN', 'MINOR_CN', 'NORMAL_CN', 'TUMOR_PURITY' ]\n",
    "bed_df = pd.DataFrame (columns = colnames) \n",
    "\n",
    "print (FACETCNV_OUTPUT_PATH)\n",
    "\n",
    "for line in input_file.readlines():\n",
    "    if \"gz\" in FACETCNV_OUTPUT_PATH:\n",
    "\t    line = line.decode('utf-8')      # ← gzip file일 경우\n",
    "    line = line.rstrip(\"\\n\")\n",
    "    \n",
    "    if line[0] == \"#\": # Header 저장\n",
    "        if line[0:8] == \"##purity\":\n",
    "            TUMOR_PURITY = line.split(\"=\")[1]\n",
    "            if TUMOR_PURITY == \"NA\":\n",
    "                TUMOR_PURITY = \"1.0\"\n",
    "        elif line[0:8] == \"##ploidy\":\n",
    "            TOTAL_PLOIDY = line.split(\"=\")[1]\n",
    "        continue\n",
    "    \n",
    "    else:\n",
    "        line = line.split(\"\\t\")\n",
    "        CHR, POS, REF, ALT , info_list, info_dict = parsing(line)\n",
    "        START, END = POS, info_dict[\"END\"]\n",
    "        try:   #\"LCN_EM에 . 같은 게 있어서 에러날 때가 있다\"\n",
    "            MAJOR_CN, MINOR_CN =  int(info_dict[\"TCN_EM\"])  - int (info_dict[\"LCN_EM\"]),  int (info_dict[\"LCN_EM\"])\n",
    "            NORMAL_CN = 2           \n",
    "        except:\n",
    "            continue\n",
    "        output_line = [str(CHR), str(START), str(END), str(MAJOR_CN), str(MINOR_CN), str(NORMAL_CN),  str(TUMOR_PURITY) ]\n",
    "        output_dict = {}\n",
    "        for colname_index, colname in enumerate (colnames):\n",
    "            output_dict[colname] = output_line[colname_index]\n",
    "            \n",
    "        bed_df = bed_df.append ( pd.Series( output_dict ), ignore_index = True )\n",
    "\n",
    "\n",
    "bed_df.tail(15)\n",
    "\n",
    "bed_df.to_csv (FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".bed_df\", sep = \"\\t\", index = False, header = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>MAJOR_CN</th>\n",
       "      <th>MINOR_CN</th>\n",
       "      <th>NORMAL_CN</th>\n",
       "      <th>TUMOR_PURITY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65975</td>\n",
       "      <td>53050524</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>53062325</td>\n",
       "      <td>112541963</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>112550632</td>\n",
       "      <td>115033619</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>115033817</td>\n",
       "      <td>149943989</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149944187</td>\n",
       "      <td>151707487</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>chr20</td>\n",
       "      <td>6214678</td>\n",
       "      <td>25294209</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>chr20</td>\n",
       "      <td>25294986</td>\n",
       "      <td>64327986</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>chr21</td>\n",
       "      <td>5031972</td>\n",
       "      <td>46664375</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>chr22</td>\n",
       "      <td>10685675</td>\n",
       "      <td>50782266</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>chrX</td>\n",
       "      <td>284272</td>\n",
       "      <td>156004380</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHR      START        END MAJOR_CN MINOR_CN NORMAL_CN TUMOR_PURITY\n",
       "0    chr1      65975   53050524        0        0         2          1.0\n",
       "1    chr1   53062325  112541963        1        1         2          1.0\n",
       "2    chr1  112550632  115033619        1        0         2          1.0\n",
       "3    chr1  115033817  149943989        1        1         2          1.0\n",
       "4    chr1  149944187  151707487        1        0         2          1.0\n",
       "..    ...        ...        ...      ...      ...       ...          ...\n",
       "85  chr20    6214678   25294209        2        1         2          1.0\n",
       "86  chr20   25294986   64327986        1        1         2          1.0\n",
       "87  chr21    5031972   46664375        1        1         2          1.0\n",
       "88  chr22   10685675   50782266        0        0         2          1.0\n",
       "89   chrX     284272  156004380        1        0         2          1.0\n",
       "\n",
       "[90 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "########## Mutectt2  ###############\n",
      "chr1_2787605\t220930_Dura\t190\t5\t2\t0\t0\t1.0\tMutect2\n",
      "chr1_42154829\t220930_Dura\t186\t5\t2\t0\t0\t1.0\tMutect2\n",
      "chr1_44802872\t220930_Dura\t98\t14\t2\t0\t0\t1.0\tMutect2\n",
      "chr1_243221776\t220930_Dura\t492\t48\t2\t1\t1\t1.0\tMutect2\n",
      "chr2_3644016\t220930_Dura\t830\t63\t2\t1\t1\t1.0\tMutect2\n",
      "chr2_44706287\t220930_Dura\t278\t16\t2\t1\t1\t1.0\tMutect2\n",
      "chr2_178554137\t220930_Dura\t603\t19\t2\t2\t2\t1.0\tMutect2\n",
      "chr2_185808060\t220930_Dura\t113\t5\t2\t2\t1\t1.0\tMutect2\n",
      "chr2_222909077\t220930_Dura\t85\t5\t2\t1\t1\t1.0\tMutect2\n",
      "chr2_241240049\t220930_Dura\t187\t5\t2\t1\t1\t1.0\tMutect2\n",
      "chr3_11817144\t220930_Dura\t363\t20\t2\t1\t1\t1.0\tMutect2\n",
      "chr3_49721007\t220930_Dura\t397\t25\t2\t1\t1\t1.0\tMutect2\n",
      "chr5_71656712\t220930_Dura\t172\t8\t2\t1\t1\t1.0\tMutect2\n",
      "chr6_87255842\t220930_Dura\t257\t15\t2\t1\t1\t1.0\tMutect2\n",
      "chr7_106026643\t220930_Dura\t419\t24\t2\t1\t1\t1.0\tMutect2\n",
      "chr7_122386320\t220930_Dura\t39\t4\t2\t1\t1\t1.0\tMutect2\n",
      "chr8_10622949\t220930_Dura\t168\t11\t2\t1\t1\t1.0\tMutect2\n",
      "chr8_81801465\t220930_Dura\t121\t5\t2\t1\t1\t1.0\tMutect2\n",
      "chr8_85126674\t220930_Dura\t323\t21\t2\t1\t1\t1.0\tMutect2\n",
      "chr9_127673284\t220930_Dura\t107\t0\t2\t1\t0\t1.0\tMutect2\n",
      "chr10_34360246\t220930_Dura\t262\t11\t2\t1\t1\t1.0\tMutect2\n",
      "chr15_40252764\t220930_Dura\t193\t12\t2\t1\t1\t1.0\tMutect2\n",
      "chr15_43524923\t220930_Dura\t308\t22\t2\t1\t1\t1.0\tMutect2\n",
      "chr16_3679864\t220930_Dura\t282\t24\t2\t0\t0\t1.0\tMutect2\n",
      "chr17_38361755\t220930_Dura\t255\t27\t2\t1\t1\t1.0\tMutect2\n",
      "chr18_54884282\t220930_Dura\t309\t21\t2\t1\t1\t1.0\tMutect2\n",
      "chr18_74559301\t220930_Dura\t147\t5\t2\t1\t1\t1.0\tMutect2\n",
      "chr20_13849480\t220930_Dura\t221\t7\t2\t2\t1\t1.0\tMutect2\n",
      "chr22_29604113\t220930_Dura\t298\t21\t2\t0\t0\t1.0\tMutect2\n"
     ]
    }
   ],
   "source": [
    "\n",
    "######## Mutect call을 골라줌 ###################\n",
    "bed_pybed_object = pybedtools.BedTool.from_dataframe(bed_df)                   # CNV 정보를 가지고 있는 segment 정보\n",
    "# mutect_pybed_object = pybedtools.BedTool( MUTECT_OUTPUT_PATH )       # mutect으로 call한 mutation\n",
    "# a = bed_pybed_object.intersect(mutect_pybed_object, wb = True) \n",
    "\n",
    "os.system (\"bedtools intersect -a \" + MUTECT_OUTPUT_PATH + \" -b \" + FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".temp1\" +  \" -wa -wb > \" + FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".temp2\")\n",
    "\n",
    "input_file = open (FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".temp2\", \"r\")\n",
    "output_file = open(FACETCNV_TO_PYCLONEVI_MATRIX_PATH, \"w\")\n",
    "\n",
    "print (\"########## Mutectt2  ###############\" )\n",
    "\n",
    "for interval in input_file.readlines():\n",
    "    interval = interval.rstrip(\"\\n\").split(\"\\t\")\n",
    "    #print (interval)\n",
    "    \n",
    "    line_dict = {}\n",
    "    \n",
    "    for index, contents in enumerate( interval ):\n",
    "        if \"GT\" in contents:\n",
    "            parsing_sample_index = index + 2\n",
    "            break\n",
    "\n",
    "    line_dict [\"mutation_id\"] = str(interval [0]) + \"_\" + str(interval[1])\n",
    "    line_dict [\"sample_id\"] = SAMPLE_ID\n",
    "    line_dict [\"ref_counts\"] = str( interval[ parsing_sample_index ].split(\":\")[1].split(\",\")[0])\n",
    "    line_dict [\"alt_counts\"] = str( interval[ parsing_sample_index ].split(\":\")[1].split(\",\")[1])\n",
    "    line_dict[\"normal_cn\"] = str (interval[-2])\n",
    "    line_dict[\"major_cn\"] = str (interval[-4])\n",
    "    line_dict[\"minor_cn\"] = str (interval[-3])\n",
    "    line_dict[\"tumour_content\"] = str( interval[-1])\n",
    "    line_dict[\"type\"] = \"Mutect2\"\n",
    "    \n",
    "    print ( '\\t'.join ( list( line_dict.values()) ) )\n",
    "    print ( '\\t'.join ( list( line_dict.values()) ), file = output_file )\n",
    "\n",
    "\n",
    "input_file.close()\n",
    "\n",
    "# os.system (\"rm -rf \" + FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".txt\")\n",
    "# os.system (\"rm -rf \" + FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".temp1\")\n",
    "# os.system (\"rm -rf \" + FACETCNV_TO_PYCLONEVI_MATRIX_PATH + \".temp2\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "########## Haplotype Caller ###############\n",
      "chr1\t27549970\t27549972\tchr1\t27549971\trs2076457\tC\tA\t6950.06\t.\tAC=2;AF=1.00;AN=2;BaseQRankSum=3.244;DB;DP=237;ExcessHet=3.0103;FS=11.577;MLEAC=2;MLEAF=1.00;MQ=60.00;MQRankSum=0.000;QD=29.96;ReadPosRankSum=0.740;SOR=0.782\tGT:AD:DP:GQ:PL\t1/1:3,229:232:99:6964,659,0\tchr1\t65975\t53050524\t0\t0\t2\t1.0\n",
      "\n",
      "chr1_27549971\t220930_Dura\t3\t229\t2\t0\t2\t1.0\tHC\n",
      "chr1\t112520502\t112520504\tchr1\t112520503\trs910697\tA\tG\t1822.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.059;DB;DP=147;ExcessHet=3.0103;FS=3.979;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.48;ReadPosRankSum=0.962;SOR=0.439\tGT:AD:DP:GQ:PL\t0/1:66,80:146:99:1830,0,1627\tchr1\t53062325\t112541963\t1\t1\t2\t1.0\n",
      "\n",
      "chr1_112520503\t220930_Dura\t66\t80\t2\t1\t2\t1.0\tHC\n",
      "chr1\t151709090\t151709092\tchr1\t151709091\trs2280474\tG\tA\t3086.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.621;DB;DP=259;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.30;ReadPosRankSum=0.051;SOR=0.696\tGT:AD:DP:GQ:PL\t0/1:128,123:251:99:3094,0,3367\tchr1\t151707731\t153419069\t2\t1\t2\t1.0\n",
      "\n",
      "chr1_151709091\t220930_Dura\t128\t123\t2\t1\t2\t1.0\tHC\n",
      "chr2\t47024329\t47024331\tchr2\t47024330\trs2304290\tG\tC\t3257.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.063;DB;DP=182;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=18.51;ReadPosRankSum=0.851;SOR=0.702\tGT:AD:DP:GQ:PL\t0/1:57,119:176:99:3265,0,1388\tchr2\t23703829\t70301814\t1\t1\t2\t1.0\n",
      "\n",
      "chr2_47024330\t220930_Dura\t57\t119\t2\t1\t2\t1.0\tHC\n",
      "chr2\t73265485\t73265487\tchr2\t73265486\trs526106\tA\tT\t3997.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.486;DB;DP=340;ExcessHet=3.0103;FS=6.500;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.34;ReadPosRankSum=0.118;SOR=0.878\tGT:AD:DP:GQ:PL\t0/1:163,161:324:99:4005,0,3854\tchr2\t70449565\t136544837\t1\t1\t2\t1.0\n",
      "\n",
      "chr2_73265486\t220930_Dura\t163\t161\t2\t1\t2\t1.0\tHC\n",
      "chr3\t10844234\t10844236\tchr3\t10844235\trs2304725\tT\tC\t2213.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.045;DB;DP=204;ExcessHet=3.0103;FS=1.806;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.29;ReadPosRankSum=0.028;SOR=0.832\tGT:AD:DP:GQ:PL\t0/1:102,94:196:99:2221,0,2585\tchr3\t197640\t47008337\t1\t1\t2\t1.0\n",
      "\n",
      "chr3_10844235\t220930_Dura\t102\t94\t2\t1\t2\t1.0\tHC\n",
      "chr3\t58429862\t58429864\tchr3\t58429863\trs2070662\tC\tA\t3833.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.916;DB;DP=266;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=14.86;ReadPosRankSum=-0.044;SOR=0.706\tGT:AD:DP:GQ:PL\t0/1:115,143:258:99:3841,0,2572\tchr3\t50343942\t194397836\t1\t1\t2\t1.0\n",
      "\n",
      "chr3_58429863\t220930_Dura\t115\t143\t2\t1\t2\t1.0\tHC\n",
      "chr3\t64541824\t64541826\tchr3\t64541825\trs74534712\tT\tC\t3561.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=2.770;DB;DP=274;ExcessHet=3.0103;FS=4.140;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.29;ReadPosRankSum=-0.114;SOR=0.960\tGT:AD:DP:GQ:PL\t0/1:127,141:268:99:3569,0,2814\tchr3\t50343942\t194397836\t1\t1\t2\t1.0\n",
      "\n",
      "chr3_64541825\t220930_Dura\t127\t141\t2\t1\t2\t1.0\tHC\n",
      "chr3\t123124364\t123124366\tchr3\t123124365\trs2305051\tC\tT\t2273.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.888;DB;DP=232;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=10.06;ReadPosRankSum=0.897;SOR=0.666\tGT:AD:DP:GQ:PL\t0/1:135,91:226:99:2281,0,3704\tchr3\t50343942\t194397836\t1\t1\t2\t1.0\n",
      "\n",
      "chr3_123124365\t220930_Dura\t135\t91\t2\t1\t2\t1.0\tHC\n",
      "chr3\t189907428\t189907430\tchr3\t189907429\trs1515488\tG\tA\t4270.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.686;DB;DP=363;ExcessHet=3.0103;FS=1.254;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.93;ReadPosRankSum=0.237;SOR=0.777\tGT:AD:DP:GQ:PL\t0/1:176,182:358:99:4278,0,4416\tchr3\t50343942\t194397836\t1\t1\t2\t1.0\n",
      "\n",
      "chr3_189907429\t220930_Dura\t176\t182\t2\t1\t2\t1.0\tHC\n",
      "chr4\t7939187\t7939189\tchr4\t7939188\trs7669645\tC\tT\t3729.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.661;DB;DP=406;ExcessHet=3.0103;FS=3.239;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=9.39;ReadPosRankSum=1.892;SOR=0.667\tGT:AD:DP:GQ:PL\t0/1:236,161:397:99:3737,0,5741\tchr4\t53417\t190112498\t1\t1\t2\t1.0\n",
      "\n",
      "chr4_7939188\t220930_Dura\t236\t161\t2\t1\t2\t1.0\tHC\n",
      "chr4\t38774485\t38774487\tchr4\t38774486\trs11096955\tT\tG\t2382.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=3.582;DB;DP=186;ExcessHet=3.0103;FS=2.582;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.16;ReadPosRankSum=-2.521;SOR=0.916\tGT:AD:DP:GQ:PL\t0/1:93,88:181:99:2390,0,2277\tchr4\t53417\t190112498\t1\t1\t2\t1.0\n",
      "\n",
      "chr4_38774486\t220930_Dura\t93\t88\t2\t1\t2\t1.0\tHC\n",
      "chr4\t144120554\t144120556\tchr4\t144120555\trs7687256\tT\tC\t316.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=4.207;DB;DP=227;ExcessHet=3.0103;FS=18.804;MLEAC=1;MLEAF=0.500;MQ=49.22;MQRankSum=-3.655;QD=1.43;ReadPosRankSum=-4.024;SOR=2.424\tGT:AD:DP:GQ:PL\t0/1:199,22:221:99:324,0,8252\tchr4\t53417\t190112498\t1\t1\t2\t1.0\n",
      "\n",
      "chr4_144120555\t220930_Dura\t199\t22\t2\t1\t2\t1.0\tHC\n",
      "chr4\t153788327\t153788329\tchr4\t153788328\trs3810765\tG\tA\t3428.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.292;DB;DP=276;ExcessHet=3.0103;FS=0.963;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.94;ReadPosRankSum=-0.763;SOR=0.790\tGT:AD:DP:GQ:PL\t0/1:130,135:265:99:3436,0,3410\tchr4\t53417\t190112498\t1\t1\t2\t1.0\n",
      "\n",
      "chr4_153788328\t220930_Dura\t130\t135\t2\t1\t2\t1.0\tHC\n",
      "chr5\t235432\t235434\tchr5\t235433\trs10060259\tT\tC\t2572.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.858;DB;DP=250;ExcessHet=3.0103;FS=1.015;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=10.42;ReadPosRankSum=1.412;SOR=0.767\tGT:AD:DP:GQ:PL\t0/1:141,106:247:99:2580,0,3435\tchr5\t140265\t69100902\t1\t1\t2\t1.0\n",
      "\n",
      "chr5_235433\t220930_Dura\t141\t106\t2\t1\t2\t1.0\tHC\n",
      "chr6\t32064978\t32064980\tchr6\t32064979\t.\tA\tT\t1804.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.781;DP=188;ExcessHet=3.0103;FS=2.562;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=9.70;ReadPosRankSum=-0.636;SOR=0.893\tGT:AD:DP:GQ:PL\t0/1:106,80:186:99:1812,0,2364\tchr6\t17115164\t45321820\t1\t1\t2\t1.0\n",
      "\n",
      "chr6_32064979\t220930_Dura\t106\t80\t2\t1\t2\t1.0\tHC\n",
      "chr6\t44176192\t44176194\tchr6\t44176193\trs142573641\tA\tAGGAGGAGAACGTCTCCCTGACC\t3269.60\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=2.045;DB;DP=333;ExcessHet=3.0103;FS=7.025;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.34;ReadPosRankSum=-1.772;SOR=0.837\tGT:AD:DP:GQ:PL\t0/1:173,92:265:99:3277,0,6941\tchr6\t17115164\t45321820\t1\t1\t2\t1.0\n",
      "\n",
      "chr6_44176193\t220930_Dura\t173\t92\t2\t1\t2\t1.0\tHC\n",
      "chr6\t167341213\t167341215\tchr6\t167341214\trs909546\tC\tT\t4488.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=3.802;DB;DP=366;ExcessHet=3.0103;FS=0.803;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.72;ReadPosRankSum=0.925;SOR=0.757\tGT:AD:DP:GQ:PL\t0/1:184,169:353:99:4496,0,4246\tchr6\t45322806\t170584467\t1\t1\t2\t1.0\n",
      "\n",
      "chr6_167341214\t220930_Dura\t184\t169\t2\t1\t2\t1.0\tHC\n",
      "chr7\t48198204\t48198206\tchr7\t48198205\trs2188\tC\tT\t1583.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.707;DB;DP=129;ExcessHet=3.0103;FS=1.721;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.54;ReadPosRankSum=2.253;SOR=0.906\tGT:AD:DP:GQ:PL\t0/1:60,57:117:99:1591,0,1684\tchr7\t7067021\t55497585\t1\t1\t2\t1.0\n",
      "\n",
      "chr7_48198205\t220930_Dura\t60\t57\t2\t1\t2\t1.0\tHC\n",
      "chr7\t81959319\t81959321\tchr7\t81959320\trs1229502\tG\tA\t5171.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=2.672;DB;DP=442;ExcessHet=3.0103;FS=0.518;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.86;ReadPosRankSum=-0.430;SOR=0.737\tGT:AD:DP:GQ:PL\t0/1:225,211:436:99:5179,0,5187\tchr7\t76995481\t98212577\t2\t1\t2\t1.0\n",
      "\n",
      "chr7_81959320\t220930_Dura\t225\t211\t2\t1\t2\t1.0\tHC\n",
      "chr7\t123624749\t123624751\tchr7\t123624750\trs11769381\tC\tT\t3624.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.617;DB;DP=273;ExcessHet=3.0103;FS=3.435;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.47;ReadPosRankSum=-1.126;SOR=0.912\tGT:AD:DP:GQ:PL\t0/1:125,144:269:99:3632,0,3170\tchr7\t105610458\t142520619\t1\t1\t2\t1.0\n",
      "\n",
      "chr7_123624750\t220930_Dura\t125\t144\t2\t1\t2\t1.0\tHC\n",
      "chr7\t141764964\t141764966\tchr7\t141764965\trs2270009\tC\tT\t3447.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.365;DB;DP=403;ExcessHet=3.0103;FS=0.374;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=8.86;ReadPosRankSum=-0.762;SOR=0.662\tGT:AD:DP:GQ:PL\t0/1:231,158:389:99:3455,0,5510\tchr7\t105610458\t142520619\t1\t1\t2\t1.0\n",
      "\n",
      "chr7_141764965\t220930_Dura\t231\t158\t2\t1\t2\t1.0\tHC\n",
      "chr7\t142753013\t142753015\tchr7\t142753014\trs6667\tT\tC\t3796.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.630;DB;DP=315;ExcessHet=3.0103;FS=0.871;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.49;ReadPosRankSum=-0.105;SOR=0.765\tGT:AD:DP:GQ:PL\t0/1:142,162:304:99:3804,0,3170\tchr7\t142619530\t159260919\t2\t1\t2\t1.0\n",
      "\n",
      "chr7_142753014\t220930_Dura\t142\t162\t2\t1\t2\t1.0\tHC\n",
      "chr8\t494484\t494486\tchr8\t494485\trs6999653\tA\tG\t4867.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.177;DB;DP=532;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=9.54;ReadPosRankSum=-0.031;SOR=0.695\tGT:AD:DP:GQ:PL\t0/1:289,221:510:99:4875,0,6716\tchr8\t232808\t48729906\t1\t1\t2\t1.0\n",
      "\n",
      "chr8_494485\t220930_Dura\t289\t221\t2\t1\t2\t1.0\tHC\n",
      "chr8\t18400592\t18400594\tchr8\t18400593\trs1799930\tG\tA\t2440.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=2.860;DB;DP=317;ExcessHet=3.0103;FS=0.928;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=8.03;ReadPosRankSum=1.266;SOR=0.768\tGT:AD:DP:GQ:PL\t0/1:198,106:304:99:2448,0,4610\tchr8\t232808\t48729906\t1\t1\t2\t1.0\n",
      "\n",
      "chr8_18400593\t220930_Dura\t198\t106\t2\t1\t2\t1.0\tHC\n",
      "chr9\t5969306\t5969308\tchr9\t5969307\trs73383144\tC\tT\t1992.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.556;DB;DP=207;ExcessHet=3.0103;FS=3.227;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=9.91;ReadPosRankSum=1.216;SOR=0.938\tGT:AD:DP:GQ:PL\t0/1:116,85:201:99:2000,0,3072\tchr9\t117807\t124540342\t1\t1\t2\t1.0\n",
      "\n",
      "chr9_5969307\t220930_Dura\t116\t85\t2\t1\t2\t1.0\tHC\n",
      "chr10\t101550816\t101550818\tchr10\t101550817\t.\tC\tA\t2079.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.463;DP=205;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=10.61;ReadPosRankSum=1.780;SOR=0.707\tGT:AD:DP:GQ:PL\t0/1:105,91:196:99:2087,0,2543\tchr10\t72005865\t103897603\t1\t1\t2\t1.0\n",
      "\n",
      "chr10_101550817\t220930_Dura\t105\t91\t2\t1\t2\t1.0\tHC\n",
      "chr11\t1190168\t1190170\tchr11\t1190169\trs1227347422\tG\tA\t3545.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=4.753;DB;DP=614;ExcessHet=3.0103;FS=2.853;MLEAC=1;MLEAF=0.500;MQ=50.92;MQRankSum=-8.994;QD=6.00;ReadPosRankSum=-0.768;SOR=0.941\tGT:AD:DP:GQ:PL\t0/1:426,165:591:99:3553,0,10687\tchr11\t1009417\t46512882\t1\t1\t2\t1.0\n",
      "\n",
      "chr11_1190169\t220930_Dura\t426\t165\t2\t1\t2\t1.0\tHC\n",
      "chr11\t1836039\t1836041\tchr11\t1836040\trs907609\tC\tT\t4346.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.996;DB;DP=385;ExcessHet=3.0103;FS=1.205;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.41;ReadPosRankSum=-2.010;SOR=0.625\tGT:AD:DP:GQ:PL\t0/1:192,189:381:99:4354,0,4595\tchr11\t1009417\t46512882\t1\t1\t2\t1.0\n",
      "\n",
      "chr11_1836040\t220930_Dura\t192\t189\t2\t1\t2\t1.0\tHC\n",
      "chr11\t1840937\t1840939\tchr11\t1840938\trs2271442\tC\tT\t4285.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.595;DB;DP=395;ExcessHet=3.0103;FS=9.110;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.37;ReadPosRankSum=1.016;SOR=1.092\tGT:AD:DP:GQ:PL\t0/1:201,176:377:99:4293,0,4969\tchr11\t1009417\t46512882\t1\t1\t2\t1.0\n",
      "\n",
      "chr11_1840938\t220930_Dura\t201\t176\t2\t1\t2\t1.0\tHC\n",
      "chr11\t119336228\t119336230\tchr11\t119336229\t.\tA\tG\t4294.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.307;DP=384;ExcessHet=3.0103;FS=2.172;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.33;ReadPosRankSum=-0.700;SOR=0.826\tGT:AD:DP:GQ:PL\t0/1:203,176:379:99:4302,0,5158\tchr11\t68911586\t134986813\t1\t1\t2\t1.0\n",
      "\n",
      "chr11_119336229\t220930_Dura\t203\t176\t2\t1\t2\t1.0\tHC\n",
      "chr11\t128805505\t128805507\tchr11\t128805506\trs7127621\tC\tT\t2905.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.581;DB;DP=284;ExcessHet=3.0103;FS=0.928;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=10.34;ReadPosRankSum=1.386;SOR=0.776\tGT:AD:DP:GQ:PL\t0/1:153,128:281:99:2913,0,3993\tchr11\t68911586\t134986813\t1\t1\t2\t1.0\n",
      "\n",
      "chr11_128805506\t220930_Dura\t153\t128\t2\t1\t2\t1.0\tHC\n",
      "chr12\t112306038\t112306040\tchr12\t112306039\trs2285807\tT\tC\t1771.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.585;DB;DP=132;ExcessHet=3.0103;FS=1.434;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.63;ReadPosRankSum=0.227;SOR=0.896\tGT:AD:DP:GQ:PL\t0/1:60,70:130:99:1779,0,1389\tchr12\t109794228\t112504927\t0\t0\t2\t1.0\n",
      "\n",
      "chr12_112306039\t220930_Dura\t60\t70\t2\t0\t2\t1.0\tHC\n",
      "chr13\t102745036\t102745038\tchr13\t102745037\trs7339187\tT\tG\t6945.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.276;DB;DP=611;ExcessHet=3.0103;FS=0.520;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.75;ReadPosRankSum=1.959;SOR=0.742\tGT:AD:DP:GQ:PL\t0/1:321,270:591:99:6953,0,8444\tchr13\t18177585\t114326192\t1\t1\t2\t1.0\n",
      "\n",
      "chr13_102745037\t220930_Dura\t321\t270\t2\t1\t2\t1.0\tHC\n",
      "chr13\t112880592\t112880594\tchr13\t112880593\trs9577174\tT\tC\t2303.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.455;DB;DP=254;ExcessHet=3.0103;FS=2.282;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=9.44;ReadPosRankSum=-0.484;SOR=0.873\tGT:AD:DP:GQ:PL\t0/1:149,95:244:99:2311,0,4106\tchr13\t18177585\t114326192\t1\t1\t2\t1.0\n",
      "\n",
      "chr13_112880593\t220930_Dura\t149\t95\t2\t1\t2\t1.0\tHC\n",
      "chr14\t21155130\t21155132\tchr14\t21155131\trs7145814\tT\tC\t3302.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.708;DB;DP=301;ExcessHet=3.0103;FS=0.908;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.31;ReadPosRankSum=-1.064;SOR=0.612\tGT:AD:DP:GQ:PL\t0/1:146,146:292:99:3310,0,3886\tchr14\t18340786\t106874947\t1\t0\t2\t1.0\n",
      "\n",
      "chr14_21155131\t220930_Dura\t146\t146\t2\t0\t2\t1.0\tHC\n",
      "chr14\t90699227\t90699229\tchr14\t90699228\trs1749699\tG\tA\t4827.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-2.897;DB;DP=411;ExcessHet=3.0103;FS=1.099;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.92;ReadPosRankSum=-3.804;SOR=0.624\tGT:AD:DP:GQ:PL\t0/1:188,217:405:99:4835,0,4423\tchr14\t18340786\t106874947\t1\t0\t2\t1.0\n",
      "\n",
      "chr14_90699228\t220930_Dura\t188\t217\t2\t0\t2\t1.0\tHC\n",
      "chr15\t66320626\t66320628\tchr15\t66320627\trs17851970\tT\tC\t2904.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=2.296;DB;DP=271;ExcessHet=3.0103;FS=2.707;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.13;ReadPosRankSum=0.101;SOR=0.874\tGT:AD:DP:GQ:PL\t0/1:131,130:261:99:2912,0,2778\tchr15\t41762451\t101849591\t1\t1\t2\t1.0\n",
      "\n",
      "chr15_66320627\t220930_Dura\t131\t130\t2\t1\t2\t1.0\tHC\n",
      "chr15\t81117764\t81117766\tchr15\t81117765\trs1533055\tG\tA\t6674.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.764;DB;DP=511;ExcessHet=3.0103;FS=0.000;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.30;ReadPosRankSum=0.258;SOR=0.689\tGT:AD:DP:GQ:PL\t0/1:221,281:502:99:6682,0,4459\tchr15\t41762451\t101849591\t1\t1\t2\t1.0\n",
      "\n",
      "chr15_81117765\t220930_Dura\t221\t281\t2\t1\t2\t1.0\tHC\n",
      "chr16\t655919\t655921\tchr16\t655920\trs9930754\tG\tC\t2444.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.917;DB;DP=197;ExcessHet=3.0103;FS=2.617;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=12.87;ReadPosRankSum=0.602;SOR=0.899\tGT:AD:DP:GQ:PL\t0/1:79,111:190:99:2452,0,1633\tchr16\t29584\t47453951\t0\t0\t2\t1.0\n",
      "\n",
      "chr16_655920\t220930_Dura\t79\t111\t2\t0\t2\t1.0\tHC\n",
      "chr17\t15717166\t15717168\tchr17\t15717167\trs2530064\tC\tT\t2893.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.100;DB;DP=354;ExcessHet=3.0103;FS=21.470;MLEAC=1;MLEAF=0.500;MQ=53.08;MQRankSum=-7.850;QD=8.51;ReadPosRankSum=-3.510;SOR=1.375\tGT:AD:DP:GQ:PL\t0/1:210,130:340:99:2901,0,5393\tchr17\t7846304\t83135891\t1\t1\t2\t1.0\n",
      "\n",
      "chr17_15717167\t220930_Dura\t210\t130\t2\t1\t2\t1.0\tHC\n",
      "chr18\t7034508\t7034510\tchr18\t7034509\trs566655\tT\tG\t3628.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=4.703;DB;DP=287;ExcessHet=3.0103;FS=8.456;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=13.01;ReadPosRankSum=1.175;SOR=1.140\tGT:AD:DP:GQ:PL\t0/1:141,138:279:99:3636,0,3082\tchr18\t48126\t80247162\t1\t1\t2\t1.0\n",
      "\n",
      "chr18_7034509\t220930_Dura\t141\t138\t2\t1\t2\t1.0\tHC\n",
      "chr19\t19711633\t19711635\tchr19\t19711634\trs74349485\tA\tG\t2080.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.915;DB;DP=291;ExcessHet=3.0103;FS=0.931;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=7.25;ReadPosRankSum=1.100;SOR=0.620\tGT:AD:DP:GQ:PL\t0/1:181,106:287:99:2088,0,4456\tchr19\t19258461\t44778297\t0\t0\t2\t1.0\n",
      "\n",
      "chr19_19711634\t220930_Dura\t181\t106\t2\t0\t2\t1.0\tHC\n",
      "chr19\t35654525\t35654527\tchr19\t35654526\trs3761091\tG\tA\t642.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-1.232;DB;DP=66;ExcessHet=3.0103;FS=5.201;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=10.37;ReadPosRankSum=0.772;SOR=0.283\tGT:AD:DP:GQ:PL\t0/1:34,28:62:99:650,0,898\tchr19\t19258461\t44778297\t0\t0\t2\t1.0\n",
      "\n",
      "chr19_35654526\t220930_Dura\t34\t28\t2\t0\t2\t1.0\tHC\n",
      "chr21\t10413733\t10413735\tchr21\t10413734\trs73151874\tA\tG\t4570.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.996;DB;DP=178;ExcessHet=3.0103;FS=3.823;MLEAC=1;MLEAF=0.500;MQ=58.56;MQRankSum=-2.929;QD=26.89;ReadPosRankSum=1.318;SOR=0.637\tGT:AD:DP:GQ:PL\t0/1:56,114:170:99:4578,0,2008\tchr21\t5031972\t46664375\t1\t1\t2\t1.0\n",
      "\n",
      "chr21_10413734\t220930_Dura\t56\t114\t2\t1\t2\t1.0\tHC\n",
      "chr21\t44628007\t44628009\tchr21\t44628008\trs9979457\tG\tT\t3540.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=1.299;DB;DP=332;ExcessHet=3.0103;FS=1.332;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=11.06;ReadPosRankSum=-2.033;SOR=0.797\tGT:AD:DP:GQ:PL\t0/1:172,148:320:99:3548,0,4016\tchr21\t5031972\t46664375\t1\t1\t2\t1.0\n",
      "\n",
      "chr21_44628008\t220930_Dura\t172\t148\t2\t1\t2\t1.0\tHC\n",
      "chr22\t19962711\t19962713\tchr22\t19962712\trs4633\tC\tT\t2073.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=0.478;DB;DP=346;ExcessHet=3.0103;FS=5.373;MLEAC=1;MLEAF=0.500;MQ=60.00;MQRankSum=0.000;QD=6.17;ReadPosRankSum=1.188;SOR=0.992\tGT:AD:DP:GQ:PL\t0/1:230,106:336:99:2081,0,5478\tchr22\t10685675\t50782266\t0\t0\t2\t1.0\n",
      "\n",
      "chr22_19962712\t220930_Dura\t230\t106\t2\t0\t2\t1.0\tHC\n",
      "chr22\t22376326\t22376328\tchr22\t22376327\trs79492832\tC\tT\t454.64\t.\tAC=1;AF=0.500;AN=2;BaseQRankSum=-0.085;DB;DP=221;ExcessHet=3.0103;FS=0.730;MLEAC=1;MLEAF=0.500;MQ=59.40;MQRankSum=-8.746;QD=2.08;ReadPosRankSum=2.045;SOR=0.788\tGT:AD:DP:GQ:PL\t0/1:188,31:219:99:462,0,7557\tchr22\t10685675\t50782266\t0\t0\t2\t1.0\n",
      "\n",
      "chr22_22376327\t220930_Dura\t188\t31\t2\t0\t2\t1.0\tHC\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "######## HC call을 골라줌 ###################\n",
    "print (\"\\n########## Haplotype Caller ###############\")\n",
    "\n",
    "hc_blood_pybed_object = pybedtools.BedTool ( HC_BLOOD_RANDOM_PICK_PATH )  \n",
    "hc_pybed_object = pybedtools.BedTool(HC_OUTPUT_PATH)  #VCF\n",
    "\n",
    "hc_selected_pybed_object = hc_blood_pybed_object.intersect(hc_pybed_object, wa = True, wb = True)       # Dura/Tumor에서 발견된 HC call 중에서 선택된 50개만 고름 (concordance가 95%이기 때문에 47개정도 예상)\n",
    "\n",
    "\n",
    "ab = hc_selected_pybed_object.intersect(bed_pybed_object, wa = True, wb = True)\n",
    "\n",
    "\n",
    "for index, contents in enumerate( ab[0] ):\n",
    "    if \"GT:\" in contents:\n",
    "        parsing_sample_index = index\n",
    "        break\n",
    "\n",
    "\n",
    "for interval in ab:\n",
    "    CHR, POS = interval[0], int (interval[2]) - 1\n",
    "    line_dict = {}\n",
    "\n",
    "    print (interval)\n",
    "\n",
    "    line_dict [\"mutation_id\"] = str(CHR) + \"_\"  + str(POS)\n",
    "    line_dict [\"sample_id\"] = SAMPLE_ID\n",
    "    line_dict [\"ref_counts\"] = str( interval [parsing_sample_index + 2] ).split(\":\")[1].split(\",\")[0]\n",
    "    line_dict [\"alt_counts\"] = str( interval [parsing_sample_index + 2] ).split(\":\")[1].split(\",\")[1]\n",
    "    line_dict[\"normal_cn\"] = str ( 2 )\n",
    "    line_dict[\"major_cn\"] = str ( interval[ -4 ] )\n",
    "    line_dict[\"minor_cn\"] = str ( interval[ -3 ] )\n",
    "    line_dict[\"tumour_content\"] = str( TUMOR_PURITY  )\n",
    "    line_dict[\"type\"] = \"HC\"\n",
    "    \n",
    "    print ( '\\t'.join ( list( line_dict.values()) ) )\n",
    "    print ( '\\t'.join ( list( line_dict.values()) ), file = output_file )\n",
    "   \n",
    "\n",
    "output_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9.0\n"
     ]
    }
   ],
   "source": [
    "print(pybedtools.__version__)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cnvpytor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b604463b8a902c2f122961866cb17f2bdae0c94d238d7f27f27d7b7f77c7b073"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
