{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (2016, Natrue Genetics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Locus</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SAMPLEID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MN-1025</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-1041</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-1045</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-1047</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-105</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-64448</th>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-64539</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-64636</th>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-64767</th>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MN-81</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Locus\n",
       "SAMPLEID       \n",
       "MN-1025      27\n",
       "MN-1041      24\n",
       "MN-1045      35\n",
       "MN-1047      18\n",
       "MN-105       42\n",
       "...         ...\n",
       "MN-64448     56\n",
       "MN-64539     31\n",
       "MN-64636     51\n",
       "MN-64767     25\n",
       "MN-81        20\n",
       "\n",
       "[87 rows x 1 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "input_df = pd.read_csv (\"/data/project/Meningioma/99.Meningioma_public/NatGen/41588_2016_BFng3651_MOESM63_ESM.csv\", sep = \",\" )\n",
    "NF_containing_sample = list ( input_df [input_df[\"Gene\"] == \"NF2\"] [\"SAMPLEID\"])\n",
    "#input_df = input_df [ input_df[\"SAMPLEID\"].isin(NF_containing_sample) ].reset_index(drop = True)\n",
    "pd.DataFrame ( input_df.groupby ( [\"SAMPLEID\"])[\"Locus\"].count() )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SAMPLEID</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>VCLASS</th>\n",
       "      <th>VAR_TYPE</th>\n",
       "      <th>CLR</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MN-303</td>\n",
       "      <td>1</td>\n",
       "      <td>156594176</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1:156594176_G/A</td>\n",
       "      <td>HAPLN2</td>\n",
       "      <td>ENSG00000132702</td>\n",
       "      <td>ENST00000255039</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant:deleterious</td>\n",
       "      <td>cGg/cAg</td>\n",
       "      <td>158</td>\n",
       "      <td>R/Q</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>159</td>\n",
       "      <td>578.19</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MN-303</td>\n",
       "      <td>1</td>\n",
       "      <td>179317895</td>\n",
       "      <td>ATGTGTGTGTGTGTGTGTGTGTGTG</td>\n",
       "      <td>A</td>\n",
       "      <td>1:179317895_ATGTGTGTGTGTGTGTGTGTGTGTG/A</td>\n",
       "      <td>SOAT1</td>\n",
       "      <td>ENSG00000057252</td>\n",
       "      <td>ENST00000367619</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant&amp;feature_truncation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs71751221</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>DEL</td>\n",
       "      <td>66</td>\n",
       "      <td>6629.19</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MN-303</td>\n",
       "      <td>2</td>\n",
       "      <td>173853359</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2:173853359_A/G</td>\n",
       "      <td>RAPGEF4</td>\n",
       "      <td>ENSG00000091428</td>\n",
       "      <td>ENST00000473043</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant&amp;nc_transcript_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs114667772</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20</td>\n",
       "      <td>142.24</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MN-303</td>\n",
       "      <td>5</td>\n",
       "      <td>67591131</td>\n",
       "      <td>AGACGAG</td>\n",
       "      <td>A</td>\n",
       "      <td>5:67591131_AGACGAG/A</td>\n",
       "      <td>PIK3R1</td>\n",
       "      <td>ENSG00000145675</td>\n",
       "      <td>ENST00000521381</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant&amp;feature_truncation</td>\n",
       "      <td>aaGACGAGa/aaa</td>\n",
       "      <td>575-577</td>\n",
       "      <td>KTR/K</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>DEL</td>\n",
       "      <td>349</td>\n",
       "      <td>5390.19</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MN-303</td>\n",
       "      <td>6</td>\n",
       "      <td>57466987</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>6:57466987_T/C</td>\n",
       "      <td>PRIM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>filtered.by.biotypes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>23</td>\n",
       "      <td>349.22</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2778</th>\n",
       "      <td>MN-1198</td>\n",
       "      <td>X</td>\n",
       "      <td>24225380</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>X:24225380_G/T</td>\n",
       "      <td>ZFX</td>\n",
       "      <td>ENSG00000005889</td>\n",
       "      <td>ENST00000459724</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant&amp;nc_transcript_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>26</td>\n",
       "      <td>43.20</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2779</th>\n",
       "      <td>MN-1198</td>\n",
       "      <td>X</td>\n",
       "      <td>43595373</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X:43595373_T/G</td>\n",
       "      <td>MAOA</td>\n",
       "      <td>ENSG00000189221</td>\n",
       "      <td>ENST00000338702</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>36</td>\n",
       "      <td>66.19</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2780</th>\n",
       "      <td>MN-1198</td>\n",
       "      <td>X</td>\n",
       "      <td>83141505</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X:83141505_T/G</td>\n",
       "      <td>CYLC1</td>\n",
       "      <td>ENSG00000183035</td>\n",
       "      <td>ENST00000329312</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>39</td>\n",
       "      <td>179.19</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2781</th>\n",
       "      <td>MN-1198</td>\n",
       "      <td>X</td>\n",
       "      <td>128674642</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X:128674642_T/G</td>\n",
       "      <td>OCRL</td>\n",
       "      <td>ENSG00000122126</td>\n",
       "      <td>ENST00000486673</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>intron_variant&amp;nc_transcript_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>46</td>\n",
       "      <td>49.19</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2782</th>\n",
       "      <td>MN-1198</td>\n",
       "      <td>X</td>\n",
       "      <td>128922106</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>X:128922106_T/G</td>\n",
       "      <td>SASH3</td>\n",
       "      <td>ENSG00000122122</td>\n",
       "      <td>ENST00000476532</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>non_coding_exon_variant&amp;nc_transcript_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SOMATIC</td>\n",
       "      <td>SNP</td>\n",
       "      <td>43</td>\n",
       "      <td>94.15</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2783 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SAMPLEID CHROM        POS                        REF ALT  \\\n",
       "0      MN-303     1  156594176                          G   A   \n",
       "1      MN-303     1  179317895  ATGTGTGTGTGTGTGTGTGTGTGTG   A   \n",
       "2      MN-303     2  173853359                          A   G   \n",
       "3      MN-303     5   67591131                    AGACGAG   A   \n",
       "4      MN-303     6   57466987                          T   C   \n",
       "...       ...   ...        ...                        ...  ..   \n",
       "2778  MN-1198     X   24225380                          G   T   \n",
       "2779  MN-1198     X   43595373                          T   G   \n",
       "2780  MN-1198     X   83141505                          T   G   \n",
       "2781  MN-1198     X  128674642                          T   G   \n",
       "2782  MN-1198     X  128922106                          T   G   \n",
       "\n",
       "                                        Locus     Gene           GeneID  \\\n",
       "0                             1:156594176_G/A   HAPLN2  ENSG00000132702   \n",
       "1     1:179317895_ATGTGTGTGTGTGTGTGTGTGTGTG/A    SOAT1  ENSG00000057252   \n",
       "2                             2:173853359_A/G  RAPGEF4  ENSG00000091428   \n",
       "3                        5:67591131_AGACGAG/A   PIK3R1  ENSG00000145675   \n",
       "4                              6:57466987_T/C    PRIM2              NaN   \n",
       "...                                       ...      ...              ...   \n",
       "2778                           X:24225380_G/T      ZFX  ENSG00000005889   \n",
       "2779                           X:43595373_T/G     MAOA  ENSG00000189221   \n",
       "2780                           X:83141505_T/G    CYLC1  ENSG00000183035   \n",
       "2781                          X:128674642_T/G     OCRL  ENSG00000122126   \n",
       "2782                          X:128922106_T/G    SASH3  ENSG00000122122   \n",
       "\n",
       "              Feature Feature_type  \\\n",
       "0     ENST00000255039   Transcript   \n",
       "1     ENST00000367619   Transcript   \n",
       "2     ENST00000473043   Transcript   \n",
       "3     ENST00000521381   Transcript   \n",
       "4                 NaN          NaN   \n",
       "...               ...          ...   \n",
       "2778  ENST00000459724   Transcript   \n",
       "2779  ENST00000338702   Transcript   \n",
       "2780  ENST00000329312   Transcript   \n",
       "2781  ENST00000486673   Transcript   \n",
       "2782  ENST00000476532   Transcript   \n",
       "\n",
       "                                        Consequence         Codons  \\\n",
       "0                      missense_variant:deleterious        cGg/cAg   \n",
       "1                 intron_variant&feature_truncation            NaN   \n",
       "2              intron_variant&nc_transcript_variant            NaN   \n",
       "3               missense_variant&feature_truncation  aaGACGAGa/aaa   \n",
       "4                              filtered.by.biotypes            NaN   \n",
       "...                                             ...            ...   \n",
       "2778           intron_variant&nc_transcript_variant            NaN   \n",
       "2779                                 intron_variant            NaN   \n",
       "2780                                 intron_variant            NaN   \n",
       "2781           intron_variant&nc_transcript_variant            NaN   \n",
       "2782  non_coding_exon_variant&nc_transcript_variant            NaN   \n",
       "\n",
       "     Protein_position Amino_acids Existing_variation   VCLASS VAR_TYPE  CLR  \\\n",
       "0                 158         R/Q                NaN  SOMATIC      SNP  159   \n",
       "1                 NaN         NaN         rs71751221  SOMATIC      DEL   66   \n",
       "2                 NaN         NaN        rs114667772  SOMATIC      SNP   20   \n",
       "3             575-577       KTR/K                NaN  SOMATIC      DEL  349   \n",
       "4                 NaN         NaN                NaN  SOMATIC      SNP   23   \n",
       "...               ...         ...                ...      ...      ...  ...   \n",
       "2778              NaN         NaN                NaN  SOMATIC      SNP   26   \n",
       "2779              NaN         NaN                NaN  SOMATIC      SNP   36   \n",
       "2780              NaN         NaN                NaN  SOMATIC      SNP   39   \n",
       "2781              NaN         NaN                NaN  SOMATIC      SNP   46   \n",
       "2782              NaN         NaN                NaN  SOMATIC      SNP   43   \n",
       "\n",
       "         QUAL FILTER  \n",
       "0      578.19   PASS  \n",
       "1     6629.19   PASS  \n",
       "2      142.24   PASS  \n",
       "3     5390.19   PASS  \n",
       "4      349.22   PASS  \n",
       "...       ...    ...  \n",
       "2778    43.20   PASS  \n",
       "2779    66.19   PASS  \n",
       "2780   179.19   PASS  \n",
       "2781    49.19   PASS  \n",
       "2782    94.15   PASS  \n",
       "\n",
       "[2783 rows x 20 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 9.1 seconds.\n",
      "Starting matrix generation for INDELs...Completed! Elapsed time: 3.34 seconds.\n",
      "Matrices generated for 1 samples with 189 errors. Total of 2092 SNVs, 3 DINUCs, and 502 INDELs were successfully analyzed.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.system (\"rm -rf /data/project/Meningioma/99.Meningioma_public/NatGen/41.Signature/01.SigProfiler/11.matrix/*\")\n",
    "\n",
    "SigProfMatrix_Total = pd.DataFrame (  columns = [\"Project\", \"Sample\", \"ID\", \"Genome\", \"mut_type\", \"chrom\",\t\"pos_start\",\"pos_end\",\t\"ref\",\t\"alt\"\t,\"Type\"] ) \n",
    "SigProfMatrix_Total[\"ID\"] = input_df [\"SAMPLEID\"] + \"_\" + input_df[\"Locus\"]\n",
    "SigProfMatrix_Total[\"Genome\"] = \"GRCh37\"\n",
    "SigProfMatrix_Total[\"mut_type\"] = \"SNP\"\n",
    "SigProfMatrix_Total[\"chrom\"] = input_df[\"CHROM\"]\n",
    "SigProfMatrix_Total[\"pos_start\"] = input_df[\"POS\"]\n",
    "SigProfMatrix_Total[\"pos_end\"] = input_df[\"POS\"]\n",
    "SigProfMatrix_Total[\"ref\"] = input_df[\"REF\"]\n",
    "SigProfMatrix_Total[\"alt\"] = input_df[\"ALT\"]\n",
    "SigProfMatrix_Total[\"Type\"] = \"SOMATIC\"\n",
    "SigProfMatrix_Total [\"Project\"] = \"NatGen\"\n",
    "SigProfMatrix_Total [\"Sample\"] = \"NatGen\"\n",
    "SigProfMatrix_Total['mut_type'] = [\n",
    "    'SNP' if len(ref) == len(alt) else 'DEL' if len(ref) > len(alt) else 'IND'\n",
    "    for ref, alt in zip(SigProfMatrix_Total['ref'], SigProfMatrix_Total['alt'])\n",
    "]\n",
    "SigProfMatrix_Total.to_csv ( \"/data/project/Meningioma/99.Meningioma_public/NatGen/41.Signature/01.SigProfiler/11.matrix/Variants.txt\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "# Matrix Generator\n",
    "from SigProfilerMatrixGenerator.scripts import SigProfilerMatrixGeneratorFunc as matGen\n",
    "\n",
    "\n",
    "matrices = matGen.SigProfilerMatrixGeneratorFunc ( \"NatGen\",  # project\n",
    "                                                                                    \"GRCh37\",   # reference_genome\n",
    "                                                                                    \"/data/project/Meningioma/99.Meningioma_public/NatGen/41.Signature/01.SigProfiler/11.matrix/\",  # path_to_input_files \n",
    "                                                                                    exome=False, \n",
    "                                                                                    bed_file=None, \n",
    "                                                                                    chrom_based=False, \n",
    "                                                                                    plot=True, \n",
    "                                                                                    tsb_stat=False, \n",
    "                                                                                    seqInfo=True)\n",
    "                                                                                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The following signatures are excluded: SBS11 SBS25 SBS31 SBS35 SBS86 SBS87 SBS90 SBS11 SBS25 SBS31 SBS32 SBS35 SBS86 SBS87 SBS90 SBS27 SBS43 SBS45 SBS46 SBS47 SBS48 SBS49 SBS50 SBS51 SBS52 SBS53 SBS54 SBS55 SBS56 SBS57 SBS58 SBS59 SBS60 SBS95 SBS8 SBS12 SBS16 SBS17a SBS17b SBS19 SBS23 SBS28 SBS33 SBS34 SBS37 SBS39 SBS40 SBS41 SBS89 SBS91 SBS93 SBS94\n",
      "Assigning COSMIC sigs or Signature Database ...... \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|████████████████████████████████████████| 1/1 [100%] in 0.1s (11.92/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "from SigProfilerAssignment import Analyzer as Analyze\n",
    "\n",
    "Analyze.cosmic_fit (\"/data/project/Meningioma/99.Meningioma_public/NatGen/41.Signature/01.SigProfiler/11.matrix/output/SBS/NatGen.SBS96.all\", \n",
    "                                \"/data/project/Meningioma/99.Meningioma_public/NatGen/41.Signature/01.SigProfiler/11.matrix/output/Assignment\", \n",
    "                                input_type=\"matrix\", \n",
    "                                cosmic_version=3.3, \n",
    "                                exome=False,\n",
    "                                genome_build=\"GRCh37\", \n",
    "                                signature_database=None,\n",
    "                                exclude_signature_subgroups = [ \"Chemotherapy_signatures\", \"Treatment_signatures\", \"Artifact_signatures\", \"Unknown_signatures\", \"Test2\" ],   #  'remove' 가 붙은건 cnvpytor, 없는건 master\n",
    "                                #exclude_signature_subgroups = [ 'Moore_signatures'],   #  'remove' 가 붙은건 cnvpytor, 없는건 master\n",
    "                                export_probabilities=False,\n",
    "                                export_probabilities_per_mutation=False, \n",
    "                                make_plots=True,\n",
    "                                verbose=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
