{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Multiple Mutect에서 shared call 구하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MIN_BLOCK_SHARED = 1\n",
      "\n",
      "\n",
      "/home/goldpm1/Meningioma/04.mutect/03.vep/220930_multiple.MT2.FMC.HF.RMBLACK.vep.vcf\n",
      "CHR:POS\tGene\tRef\tAlt\tVariant_Classification\t220930_Dura\t220930_Tumor\n",
      "chr1:44802872\tTCTEX1D4\tA\tC\tdownstream_gene_variant\t98.14\t26.13\n",
      "chr1:150487535\tTARS2\tA\tG\tintron_variant&NMD_transcript_variant\t124.8\t64.26\n",
      "chr1:243221776\tCEP170\tT\tC\tmissense_variant\t492.48\t224.115\n",
      "chr2:3644016\tCOLEC11\tC\tT\tsynonymous_variant\t830.63\t440.139\n",
      "chr2:44706287\tCAMKMT\tG\tC\tmissense_variant&splice_region_variant\t278.16\t140.39\n",
      "chr2:178554137\tTTN\tA\tT\tsynonymous_variant\t603.19\t323.84\n",
      "chr3:11817144\tTAMM41\tT\tA\tintron_variant\t363.20\t238.91\n",
      "chr3:49721007\tGMPPB\tC\tT\t3_prime_UTR_variant\t397.25\t241.83\n",
      "chr3:50194058\tSEMA3F\tC\tA\tdownstream_gene_variant\t150.8\t72.36\n",
      "chr3:52322739\tDNAH1\tC\tT\tsynonymous_variant\t476.2\t338.21\n",
      "chr5:71656712\tMCCC2\tA\tT\tintron_variant\t174.8\t88.28\n",
      "chr5:103108723\tGIN1\tA\tG\tintron_variant\t211.11\t127.32\n",
      "chr6:32554712\tHLA-DRB6\tC\tT\tnon_coding_transcript_exon_variant\t76.8\t51.2\n",
      "chr6:87255842\tZNF292\tA\tG\tmissense_variant\t257.15\t102.50\n",
      "chr7:106026643\tCDHR3\tTTTC\tT\tintron_variant\t419.24\t277.110\n",
      "chr8:10622949\tRP1L1\tG\tC\tnon_coding_transcript_exon_variant\t168.11\t40.39\n",
      "chr8:81801465\tSNX16\tT\tC\t3_prime_UTR_variant\t121.5\t36.23\n",
      "chr8:85126674\tLRRCC1\tG\tC\tintron_variant\t323.21\t66.77\n",
      "chr10:34360246\tPARD3\tT\tA\tmissense_variant\t262.11\t109.75\n",
      "chr12:57730658\tAGAP2\tC\tT\tintron_variant\t45.5\t21.16\n",
      "chr12:106860433\tRIC8B\tT\tG\tintron_variant\t62.4\t41.18\n",
      "chr12:131165351\tLINC01257\tG\tA\tintron_variant&non_coding_transcript_variant\t149.3\t62.23\n",
      "chr13:28438321\tFLT1\tA\tG\tmissense_variant\t239.7\t149.10\n",
      "chr13:113623173\tTFDP1\tG\tA\tsplice_region_variant&intron_variant\t131.5\t52.12\n",
      "chr15:40252764\tPAK6\tG\tA\tintron_variant\t195.12\t112.46\n",
      "chr15:43524923\tMAP1A\tA\tC\tsynonymous_variant\t308.22\t187.64\n",
      "chr15:43876007\tFRMD5\tC\tT\tintron_variant\t46.3\t24.6\n",
      "chr15:48332213\tDUT\tT\tG\tintron_variant\t230.17\t110.56\n",
      "chr15:78484821\tIREB2\tT\tTA\tstop_gained&frameshift_variant\t162.5\t77.17\n",
      "chr16:2235405\tE4F1\tG\tA\tmissense_variant\t292.4\t220.3\n",
      "chr16:3679864\tTRAP1\tG\tA\tintron_variant\t281.24\t98.60\n",
      "chr17:7254728\tCTDNEP1\tT\tC\tupstream_gene_variant\t358.7\t207.22\n",
      "chr17:38361755\tSOCS7\tC\tG\tmissense_variant\t255.27\t133.57\n",
      "chr17:76478828\tRHBDF2\tT\tC\tmissense_variant\t116.5\t67.16\n",
      "chr18:54884282\tRAB27B\tC\tT\tintron_variant\t309.21\t130.74\n",
      "chr19:55341955\tKMT5C\tA\tG\tmissense_variant\t529.9\t338.25\n",
      "chr21:25741676\tGABPA\tGT\tG\tsplice_donor_variant\t63.2\t28.12\n",
      "chr22:29604113\tNF2\tG\tC\tsplice_donor_variant\t298.21\t81.77\n",
      "총 Variant = 38개\n",
      "\n",
      "\n",
      "/home/goldpm1/Meningioma/04.mutect/03.vep/221026_multiple.MT2.FMC.HF.RMBLACK.vep.vcf\n",
      "CHR:POS\tGene\tRef\tAlt\tVariant_Classification\t221026_Dura\t221026_Tumor\n",
      "chr2:151614518\tNEB\tG\tT\tmissense_variant\t261.2\t75.24\n",
      "chr2:218961004\tCDK5R2\tC\tG\t3_prime_UTR_variant\t296.2\t67.74\n",
      "chr2:219023960\tCFAP65\tG\tC\tintron_variant\t62.12\t23.18\n",
      "chr3:49058071\tQRICH1\tC\tT\tintron_variant\t152.25\t57.25\n",
      "chr3:51863268\tIQCF2\tC\tT\tsynonymous_variant\t401.65\t146.77\n",
      "chr3:194652886\tLSG1\tCT\tC\tframeshift_variant\t268.61\t74.46\n",
      "chr4:39302764\tRFC1\tA\tT\tmissense_variant\t237.2\t67.29\n",
      "chr4:145143908\tOTUD4\tG\tC\tintron_variant\t271.2\t87.55\n",
      "chr6:42269050\tTRERF1\tG\tC\tmissense_variant\t331.6\t128.67\n",
      "chr6:96528460\tUFL1\tC\tG\tintron_variant\t262.2\t31.49\n",
      "chr7:92371674\tANKIB1\tC\tA\tintron_variant\t103.2\t32.10\n",
      "chr7:128396595\tIMPDH1\tC\tA\tsplice_region_variant&intron_variant\t69.17\t29.14\n",
      "chr7:129206557\tSMO\tC\tT\tmissense_variant\t275.44\t109.32\n",
      "chr10:117167687\tMIR3663HG\tTGGTCCCGGCGCCCGGCCTGTGTGGTGCTCAGATCAGACTTATCCACCCGAGCACCACGCAGACCAGCGCCTGGACCAA\tT\tintron_variant&non_coding_transcript_variant\t191.28\t112.3\n",
      "chr11:5901146\tTRIM5\tC\tT\tintron_variant\t654.132\t182.96\n",
      "chr11:6545590\tDNHD1\tC\tT\tmissense_variant\t672.5\t322.3\n",
      "chr11:105023156\tCASP5\tT\tC\t5_prime_UTR_variant\t559.10\t165.111\n",
      "chr12:53251938\tMFSD5\tG\tA\tupstream_gene_variant\t284.3\t101.47\n",
      "chr12:57488053\tMARS1\tG\tC\tupstream_gene_variant\t479.2\t141.90\n",
      "chr15:34791073\tACTC1\tC\tA\tintron_variant\t198.45\t45.34\n",
      "chr15:63662022\tHERC1\tCT\tC\tsplice_acceptor_variant\t844.11\t199.98\n",
      "chr16:2707252\tKCTD5\tACCT\tA\tintron_variant\t453.3\t110.65\n",
      "chr16:31436088\tZNF843\tC\tA\tsynonymous_variant\t164.35\t50.14\n",
      "chr20:35618686\tSPAG4\tA\tC\tmissense_variant\t54.12\t28.11\n",
      "chr20:63284964\tARFGAP1\tTC\tT\tintron_variant\t368.3\t156.102\n",
      "chr21:33767858\tITSN1\tAT\tA\tintron_variant\t219.3\t46.14\n",
      "chr22:29604050\tNF2\tCA\tC\tframeshift_variant\t208.51\t67.40\n",
      "총 Variant = 27개\n",
      "\n",
      "\n",
      "/home/goldpm1/Meningioma/04.mutect/03.vep/221102_multiple.MT2.FMC.HF.RMBLACK.vep.vcf\n",
      "CHR:POS\tGene\tRef\tAlt\tVariant_Classification\t221102_Dura\t221102_Tumor\n",
      "chr1:29275768\tPTPRU\tA\tC\tintron_variant\t186.2\t88.6\n",
      "chr1:201465865\tPHLDA3\tG\tA\tdownstream_gene_variant\t188.2\t118.5\n",
      "chr1:222943374\tDISP1\tC\tT\tintron_variant\t336.3\t291.4\n",
      "chr3:115676386\tGAP43\tC\tT\tmissense_variant\t358.3\t253.4\n",
      "chr5:42719018\tGHR\tG\tA\tmissense_variant\t254.7\t171.2\n",
      "chr6:10910972\tSYCP2L\tC\tA\tintron_variant\t133.4\t98.3\n",
      "chr12:56117657\tZC3H10\tG\tA\tupstream_gene_variant\t47.4\t38.7\n",
      "chr13:45713154\tCBY2\tC\tT\tintron_variant\t412.3\t199.93\n",
      "chr17:35183515\tUNC45B\tC\tT\tmissense_variant\t133.2\t94.2\n",
      "chr20:63244559\tBIRC7\tG\tA\tdownstream_gene_variant\t166.3\t121.2\n",
      "총 Variant = 10개\n"
     ]
    }
   ],
   "source": [
    "import vcf\n",
    "import sys\n",
    "import os.path\n",
    "import argparse\n",
    "\n",
    "DIR = \"/home/goldpm1/Meningioma/04.mutect/03.vep\"\n",
    "\n",
    "MIN_BLOCK_SHARED = 1\n",
    "print (\"MIN_BLOCK_SHARED = {}\".format(MIN_BLOCK_SHARED))\n",
    "\n",
    "for INPUT_VCF in [ DIR + \"/220930_multiple.MT2.FMC.HF.RMBLACK.vep.vcf\", DIR + \"/221026_multiple.MT2.FMC.HF.RMBLACK.vep.vcf\", DIR + \"/221102_multiple.MT2.FMC.HF.RMBLACK.vep.vcf\"]:\n",
    "#for INPUT_VCF in [ DIR + \"/220930_Dura.MT2.FMC.HF.RMBLACK.vep.vcf\", DIR + \"/221026_Dura.MT2.FMC.HF.RMBLACK.vep.vcf\", DIR + \"/221102_Dura.MT2.FMC.HF.RMBLACK.vep.vcf\",  DIR + \"/220930_Tumor.MT2.FMC.HF.RMBLACK.vep.vcf\", DIR + \"/221026_Tumor.MT2.FMC.HF.RMBLACK.vep.vcf\", DIR + \"/221102_Tumor.MT2.FMC.HF.RMBLACK.vep.vcf\"]:\n",
    "    print (\"\\n\\n\" + INPUT_VCF)\n",
    "    vcf_reader = vcf.Reader(open(INPUT_VCF, \"r\"))\n",
    "    samplenames = vcf_reader.samples        # ['220930_Blood', '220930_Dura', '220930_Tumor']\n",
    "\n",
    "    Venn_data = {}\n",
    "    first_line = [\"CHR:POS\", \"Gene\", \"Ref\", \"Alt\", \"Variant_Classification\"]\n",
    "    for sampleindex, samplename in enumerate(vcf_reader.samples):\n",
    "        Venn_data[samplename] = []\n",
    "        if \"Blood\" not in samplename:\n",
    "            first_line.append (samplename)          # ['CHR', 'POS', '220930_Dura', '220930_Tumor']\n",
    "\n",
    "\n",
    "    print (*first_line, sep = \"\\t\")\n",
    "\n",
    "    variant_count_total = 0\n",
    "\n",
    "    for record in vcf_reader:        # record.CHROM, recrod.POS ,record.ALT\n",
    "        if record.CHROM == \"chrM\":\n",
    "            break\n",
    "        # variant가 몇 sample에서 나오는지 찾기\n",
    "        count = 0\n",
    "        for sampleindex, samplename in enumerate(samplenames):    # ['220930_Dura', '220930_Tumor']\n",
    "            \n",
    "            if \"Blood\" not in samplename:\n",
    "                if record.samples[sampleindex][\"GT\"] not in [\"./.\", \"0/0\", \"0|0\"] :   # Genotype 조건 만족\n",
    "                    if None not in record.samples[sampleindex][\"AD\"]:      # 왜인지는 모르겠으나 [25, 5, None] 같은 경우도 있다\n",
    "                        if record.samples[sampleindex][\"AD\"][-1] != 0:\n",
    "                            #print (record.CHROM, record.POS, samplename, record.samples[sampleindex][\"GT\"], record.samples[sampleindex][\"AD\"])\n",
    "                            count = count  + 1\n",
    "\n",
    "\n",
    "        if count >= MIN_BLOCK_SHARED:\n",
    "            try:\n",
    "                Gene_Symbol = (record.INFO[\"CSQ\"][0].split(\"|\")[3])       # 일단 맨 상위 script만 보자\n",
    "                if record.INFO[\"CSQ\"][0].split(\"|\")[1] == \"intergenic_variant\":\n",
    "                    Gene_Symbol = \"intergenic_variant\"\n",
    "                elif record.INFO[\"CSQ\"][0].split(\"|\")[1] == \"regulatory_region_variant\":\n",
    "                    Gene_Symbol = \"regulatory_region_variant\"\n",
    "            except:\n",
    "                Gene_Symbol = \"\"\n",
    "                \n",
    "            line = [str(record.CHROM) + \":\" + str(record.POS), str(Gene_Symbol), str(record.REF), str(record.ALT)[1:-1], record.INFO[\"CSQ\"][0].split(\"|\")[1] ]\n",
    "            #line = [str(record.CHROM) + \":\" + str(record.POS), str(Gene_Symbol) ]\n",
    "            \n",
    "            #print ( record.INFO[\"CSQ\"][0].split(\"|\") ) \n",
    "            \n",
    "            check = 0\n",
    "            for sampleindex, samplename in enumerate(samplenames):\n",
    "                if \"Blood\" not in samplename:\n",
    "                    line.append('.'.join([str(a) for a in record.samples[sampleindex][\"AD\"]]))       # \"REF.ALT\"\n",
    "                    if (0 in record.samples[sampleindex][\"AD\"]) | (  int ( record.samples[sampleindex][\"AD\"][1] )  < 2):\n",
    "                        check = 1\n",
    "        \n",
    "            \n",
    "            if check == 0:\n",
    "                print (\"\\t\".join(line))\n",
    "                variant_count_total = variant_count_total + 1\n",
    "\n",
    "\n",
    "    print (\"총 Variant = {}개\".format(variant_count_total))\n",
    "    # output_file.close()\n",
    "    # os.system(\"cp {0} {1}\".format(OUTPUT + \"_\" + str(MIN_BLOCK_SHARED) + \".txt\", OUTPUT2 + \"_\" + str(MIN_BLOCK_SHARED) + \".txt\"))\n",
    "\n",
    "    #Call(sample=220930_Tumor, CallData(GT=0/1, AD=[6, 0], AF=[0.143], DP=6, F1R2=[2, 0], F2R1=[3, 0], FAD=[5, 0], SB=[3, 3, 0, 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/data/project/Alzheimer/Public/2022GBM/WGS/04.Analysis/sequenza_shared/WGS_cnv_input.txt\", sep = \"\\t\")\n",
    "df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  },
  "vscode": {
   "interpreter": {
    "hash": "7b5c7a496a28c03432057dcb80d92e2013c1e0103db013fec4e486da491292b8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
