{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SigProfilerAssignment import Analyzer as Analyze\n",
    "import sigProfilerPlotting as sigPlt\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import shutil\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COSMIC DATABASE (SigProfilerAssignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 3.61 seconds.\n",
      "Starting matrix generation for INDELs...Completed! Elapsed time: 3.38 seconds.\n",
      "Matrices generated for 10 samples with 1482 errors. Total of 11849 SNVs, 216 DINUCs, and 2492 INDELs were successfully analyzed.\n",
      "Assigning COSMIC sigs or Signature Database ...... \n",
      "|████████████████████████████████████████| 10/10 [100%] in 2.7s (3.74/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerAssignment.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "input_data = '/data/project/Meningioma/41.Signature/01.SigProfiler/Tumor/01.vcf'\n",
    "output_dir_assignment = '/data/project/Meningioma/41.Signature/01.SigProfiler/Tumor/02.result_cosmic'\n",
    "output_dir_extractor = '/data/project/Meningioma/41.Signature/01.SigProfiler/Tumor/02.result_extract'\n",
    "# output_dir_assignment = '/data/project/Meningioma/41.Signature/01.SigProfiler/Tumor/02.assignment'\n",
    "# output_dir_extractor = '/data/project/Meningioma/41.Signature/01.SigProfiler/Tumor/03.extractor'\n",
    "\n",
    "if not os.path.exists(output_dir_assignment):\n",
    "        os.makedirs(output_dir_assignment)\n",
    "\n",
    "Analyze.cosmic_fit(input_data, output_dir_assignment, input_type = \"vcf\", genome_build=\"GRCh38\", context_type = \"96\", exome =True,\n",
    "                   signatures = None, signature_database = None, collapse_to_SBS96=False, make_plots = True,\n",
    "                    )  # \"Artifact_signatures\", exclude_signature_subgroups = [\"Chemotherapy_signatures\", \"Treatment_signatures\", \"Test2\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SBS signature fitting stack bar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SBS1</th>\n",
       "      <th>SBS5</th>\n",
       "      <th>SBS12</th>\n",
       "      <th>SBS15</th>\n",
       "      <th>SBS19</th>\n",
       "      <th>SBS22</th>\n",
       "      <th>SBS28</th>\n",
       "      <th>SBS37</th>\n",
       "      <th>SBS39</th>\n",
       "      <th>SBS42</th>\n",
       "      <th>SBS43</th>\n",
       "      <th>SBS54</th>\n",
       "      <th>SBS55</th>\n",
       "      <th>SBS57</th>\n",
       "      <th>SBS92</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Samples</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>220930_Tumor</th>\n",
       "      <td>7</td>\n",
       "      <td>505</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221026_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>352</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>77</td>\n",
       "      <td>172</td>\n",
       "      <td>0</td>\n",
       "      <td>150</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>167</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221102_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>177</td>\n",
       "      <td>0</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>536</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221202_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>316</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>778</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230127_Tumor</th>\n",
       "      <td>40</td>\n",
       "      <td>235</td>\n",
       "      <td>177</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>195</td>\n",
       "      <td>140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>123</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230323_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>752</td>\n",
       "      <td>267</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>162</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>147</td>\n",
       "      <td>203</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230405_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>145</td>\n",
       "      <td>0</td>\n",
       "      <td>442</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>497</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230419_Tumor</th>\n",
       "      <td>14</td>\n",
       "      <td>724</td>\n",
       "      <td>373</td>\n",
       "      <td>0</td>\n",
       "      <td>112</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>256</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>148</td>\n",
       "      <td>150</td>\n",
       "      <td>270</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230526_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>264</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230822_Tumor</th>\n",
       "      <td>0</td>\n",
       "      <td>247</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>466</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SBS1  SBS5  SBS12  SBS15  SBS19  SBS22  SBS28  SBS37  SBS39  \\\n",
       "Samples                                                                     \n",
       "220930_Tumor     7   505      0      0      0      0      0      0      0   \n",
       "221026_Tumor     0   352      0     49      0      0     77    172      0   \n",
       "221102_Tumor     0     0      0      0      0      0    177      0    459   \n",
       "221202_Tumor     0   316      0      0      0      0      0      0      0   \n",
       "230127_Tumor    40   235    177      0      0      0      0      0    195   \n",
       "230323_Tumor     0   752    267      0      0    162      0      0      0   \n",
       "230405_Tumor     0     0      0      0      0      0    145      0    442   \n",
       "230419_Tumor    14   724    373      0    112      0      0      0    256   \n",
       "230526_Tumor     0   264      0      0      0      0      0      0      0   \n",
       "230822_Tumor     0   247      0      0      0      0      0      0      0   \n",
       "\n",
       "              SBS42  SBS43  SBS54  SBS55  SBS57  SBS92  \n",
       "Samples                                                 \n",
       "220930_Tumor      0    134      0      0      0    171  \n",
       "221026_Tumor    150      0      0    167      0      0  \n",
       "221102_Tumor      0      0      0    536      0      0  \n",
       "221202_Tumor      0      0      0    778      0      0  \n",
       "230127_Tumor    140      0      0    123      0      0  \n",
       "230323_Tumor      0      0      0    147    203      0  \n",
       "230405_Tumor      0      0      0    497      0      0  \n",
       "230419_Tumor      0      0    148    150    270      0  \n",
       "230526_Tumor      0      0      0    490      0      0  \n",
       "230822_Tumor      0      0      0    466      0      0  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = open(output_dir_assignment  + '/Assignment_Solution/Activities/Assignment_Solution_Activities.txt','r')\n",
    "\n",
    "data = f.readlines()\n",
    "\n",
    "column_name = data[0].rstrip().split('\\t')\n",
    "data_split = [x.rstrip().split('\\t') for x in data[1:]]\n",
    "\n",
    "df = pd.DataFrame(data_split, columns = column_name)\n",
    "df = df.set_index(keys = 'Samples')\n",
    "\n",
    "for i in df.columns:\n",
    "    df[i] = pd.to_numeric(df[i])\n",
    "\n",
    "# 0으로 가득찬 df는 filter하기\n",
    "for column in df.columns:\n",
    "    check = set()\n",
    "    for value in df[column]:\n",
    "        check.add(value)\n",
    "    if len(check) == 1:\n",
    "        df = df.drop(column, axis = 'columns')\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Percentage 로 변환하기"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SBS5</th>\n",
       "      <th>SBS15</th>\n",
       "      <th>SBS16</th>\n",
       "      <th>SBS23</th>\n",
       "      <th>SBS38</th>\n",
       "      <th>SBS39</th>\n",
       "      <th>SBS44</th>\n",
       "      <th>SBS84</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Samples</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>220930_Dura</th>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220930_multiple</th>\n",
       "      <td>75</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221026_Dura</th>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221026_multiple</th>\n",
       "      <td>31</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221102_multiple</th>\n",
       "      <td>0</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 SBS5  SBS15  SBS16  SBS23  SBS38  SBS39  SBS44  SBS84\n",
       "Samples                                                               \n",
       "220930_Dura         0     55      0     44      0      0      0      0\n",
       "220930_multiple    75      0     24      0      0      0      0      0\n",
       "221026_Dura        49      0      0      0      0      0     20     30\n",
       "221026_multiple    31      0      0      0     21     47      0      0\n",
       "221102_multiple     0     55      0     44      0      0      0      0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SBS_name = df.columns\n",
    "total_count = {}\n",
    "\n",
    "for index, row in df.iterrows():  ## index == A.~~ row = value\n",
    "    total = 0\n",
    "    for value in row:\n",
    "        total += value\n",
    "    total_count[index] = total\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    for idx, value in enumerate(row):\n",
    "        df[SBS_name[idx]][index] = value/total_count[index]*100\n",
    "        \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots( figsize=(12, 6), nrows = 1, ncols = 1 )\n",
    "#plt.figure ( figsize = (20, 15) )\n",
    "sns.set_style(\"white\")\n",
    "ax.set_title (\"Stacked bar chart\", fontsize = 24, fontweight='heavy')\n",
    "df.plot (kind = \"bar\", stacked=True , ax = ax)\n",
    "ax.set_xticklabels (list ( df.index ), rotation  = 0, fontsize = 12, fontweight='bold')\n",
    "ax.set_xlabel(\"\")\n",
    "plt.savefig (\"/home/goldpm1/Meningioma/script/41.SigProfiler/barchart.jpg\", dpi = 300)\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DE NOVO SIGNATURE (SigProfilerExtractor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "************** Reported Current Memory Use: 2.43 GB *****************\n",
      "\n",
      "Starting matrix generation for SNVs and DINUCs...Completed! Elapsed time: 2.96 seconds.\n",
      "Starting matrix generation for INDELs...Completed! Elapsed time: 2.22 seconds.\n",
      "Matrices generated for 9 samples with 54 errors. Total of 1833 SNVs, 4 DINUCs, and 19 INDELs were successfully analyzed.\n",
      "Extracting signature 2 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 2 signatures is 22.29 seconds\n",
      "Optimization time is 3.966200590133667 seconds\n",
      "The reconstruction error is 0.028, average process stability is 0.98 and \n",
      "the minimum process stability is 0.95 for 2 signatures\n",
      "\n",
      "\n",
      "Extracting signature 3 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 3 signatures is 34.24 seconds\n",
      "Optimization time is 4.119777202606201 seconds\n",
      "The reconstruction error is 0.195, average process stability is 0.47 and \n",
      "the minimum process stability is -0.33 for 3 signatures\n",
      "\n",
      "\n",
      "Extracting signature 4 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 4 signatures is 58.46 seconds\n",
      "Optimization time is 4.0219268798828125 seconds\n",
      "The reconstruction error is 0.1714, average process stability is 0.39 and \n",
      "the minimum process stability is -0.34 for 4 signatures\n",
      "\n",
      "\n",
      "Extracting signature 5 for mutation type 96\n",
      "The matrix normalizing cutoff is 9600\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 5 signatures is 111.54 seconds\n",
      "Optimization time is 4.032492399215698 seconds\n",
      "The reconstruction error is 0.0951, average process stability is 0.27 and \n",
      "the minimum process stability is -0.63 for 5 signatures\n",
      "\n",
      "\n",
      "\n",
      " De Novo Fitting .....\n",
      "|████████████████████████████████████████| 9/9 [100%] in 0.2s (52.24/s) \n",
      "\n",
      " Decomposing De Novo Signatures  .....\n",
      "Decompositon Plot:SBS96A |▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉| 1/1 [100%] in 0.0s (15698.83/s) \n",
      "Decompositon Plot:SBS96B |▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉| 1/1 [100%] in 0.0s (13851.95/s) \n",
      "\n",
      " Assigning decomposed signature\n",
      "|████████████████████████████████████████| 9/9 [100%] in 0.3s (35.01/s) \n",
      "Extracting signature 2 for mutation type DINUC\n",
      "The matrix normalizing cutoff is 7800\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 2 signatures is 15.15 seconds\n",
      "Optimization time is 3.947788715362549 seconds\n",
      "The reconstruction error is 0.0412, average process stability is 0.91 and \n",
      "the minimum process stability is 0.81 for 2 signatures\n",
      "\n",
      "\n",
      "\n",
      " De Novo Fitting .....\n",
      "|████████████████████████████████████████| 2/2 [100%] in 0.1s (39.44/s) \n",
      "\n",
      " Decomposing De Novo Signatures  .....\n",
      "Decompositon Plot:DBS78A |▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉| 1/1 [100%] in 0.0s (14595.13/s) \n",
      "Decompositon Plot:DBS78B |▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉| 1/1 [100%] in 0.0s (17175.05/s) \n",
      "\n",
      " Assigning decomposed signature\n",
      "|████████████████████████████████████████| 2/2 [100%] in 0.1s (38.53/s) \n",
      "Extracting signature 2 for mutation type INDEL\n",
      "The matrix normalizing cutoff is 8300\n",
      "\n",
      "\n",
      "Time taken to collect 100 iterations for 2 signatures is 22.56 seconds\n",
      "Optimization time is 4.06881046295166 seconds\n",
      "The reconstruction error is 0.6368, average process stability is 0.48 and \n",
      "the minimum process stability is 0.43 for 2 signatures\n",
      "\n",
      "\n",
      "There is no signature over the thresh-hold stability. We are selecting the lowest possible number of signatures.\n",
      "\n",
      " De Novo Fitting .....\n",
      "|████████████████████████████████████████| 6/6 [100%] in 0.2s (37.69/s) \n",
      "\n",
      " Decomposing De Novo Signatures  .....\n",
      "Decompositon Plot:ID83A |▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉| 1/1 [100%] in 0.0s (15753.74/s) \n",
      "Decompositon Plot:ID83B |▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉| 1/1 [100%] in 0.0s (16609.64/s) \n",
      "\n",
      " Assigning decomposed signature\n",
      "|████████████████████████████████████████| 6/6 [100%] in 0.2s (38.63/s) \n",
      "\n",
      "\n",
      " \n",
      "Your Job Is Successfully Completed! Thank You For Using SigProfilerExtractor.\n",
      " \n"
     ]
    }
   ],
   "source": [
    "from SigProfilerExtractor import sigpro as sig\n",
    "\n",
    "input_data = '/data/project/Meningioma/41.Signature/01.SigProfiler/Dura/01.vcf'\n",
    "output_dir_extractor = '/data/project/Meningioma/41.Signature/01.SigProfiler/Dura/03.extractor'\n",
    "\n",
    "sig.sigProfilerExtractor (input_type = \"vcf\", output = output_dir_extractor, input_data = input_data, reference_genome=\"GRCh38\", context_type = \"96,DINUC,ID\", exome =True,\n",
    "                                    minimum_signatures=2, maximum_signatures=5, nmf_replicates=100, resample = True, batch_size=1, cpu=-1, gpu=False, \n",
    "                                    nmf_init=\"random\", precision= \"single\", matrix_normalization= \"gmm\", seeds = \"random\", \n",
    "                                    min_nmf_iterations= 10000, max_nmf_iterations=1000000, nmf_test_conv= 10000, nmf_tolerance= 1e-15, get_all_signature_matrices= False,\n",
    "                                    make_decomposition_plots = True)\n",
    "                   #signatures = None, signature_database = None, collapse_to_SBS96=False, make_plots = True )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cosine similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import numpy as np\n",
    "\n",
    "COSMIC_PATH = \"/home/goldpm1/resources/COSMIC_SBS/COSMIC_v3.3.1_SBS_GRCh38.txt\"\n",
    "SIGPROFILER_DENOVO_DIR=\"/data/project/Meningioma/41.Signature/01.SigProfiler/Dura/03.extractor/SBS96/All_Solutions/SBS96_3_Signatures\"\n",
    "SIGPROFILER_DENOVO_TABLE_PATH = SIGPROFILER_DENOVO_DIR + \"/Signatures/SBS96_S3_Signatures.txt\"\n",
    "\n",
    "COSMIC_df = pd.read_csv (COSMIC_PATH, sep = \"\\t\")\n",
    "SIGPROFILER_DENOVO_TABLE_df = pd.read_csv (SIGPROFILER_DENOVO_TABLE_PATH, sep = \"\\t\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SBS96A\tlst[0:5] = [0.91 0.76 0.67 0.59 0.56]\t\tlst_arg[0:5] = Index(['SBS48', 'SBS44', 'SBS37', 'SBS3', 'SBS52'], dtype='object')\n",
      "SBS96B\tlst[0:5] = [0.77 0.63 0.62 0.6  0.57]\t\tlst_arg[0:5] = Index(['SBS93', 'SBS41', 'SBS44', 'SBS3', 'SBS37'], dtype='object')\n",
      "SBS96C\tlst[0:5] = [0.8  0.72 0.63 0.54 0.47]\t\tlst_arg[0:5] = Index(['SBS14', 'SBS5', 'SBS41', 'SBS43', 'SBS22'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "# Compute cosine similarity\n",
    "cos_sim = cosine_similarity( COSMIC_df.iloc[:, 1:COSMIC_df.shape[1]].values.transpose(), SIGPROFILER_DENOVO_TABLE_df.iloc[:,1:SIGPROFILER_DENOVO_TABLE_df.shape[1]].values.transpose()  )\n",
    "\n",
    "\n",
    "# 가장 비슷한 (= cosine similarity가 가장 높은 SBS) SBS 찾기\n",
    "for signature_i in range ( cos_sim.shape[1] ):\n",
    "    lst = sorted(  cos_sim[:, signature_i], reverse=True)\n",
    "    lst_arg = []\n",
    "    for i in lst:\n",
    "        lst_arg.append ( list(cos_sim[:, signature_i]).index (i) )\n",
    "\n",
    "    max = np.max ( cos_sim[:, signature_i] )\n",
    "    max_arg = np.argmax ( cos_sim[:, signature_i] )\n",
    "    #print (\"S{}\\tmax = {}\\tmax_arg = {}\".format(signature_i + 1, round (max, 2), COSMIC_df.columns[ max_arg ] )) \n",
    "    print (\"SBS96{}\\tlst[0:5] = {}\\t\\tlst_arg[0:5] = {}\".format ( chr(signature_i + 65), np.round (lst[0:5], 2) ,  COSMIC_df.columns [ [i for i in  lst_arg [0:5]] ]  )  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(79, 96)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COSMIC_df.iloc[:, 1:COSMIC_df.shape[1]].values.transpose().shape"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cnvpytor",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  },
  "vscode": {
   "interpreter": {
    "hash": "b604463b8a902c2f122961866cb17f2bdae0c94d238d7f27f27d7b7f77c7b073"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
